---
title: "Particle associated lifestyle and genome size and depths of MAGs"
author: "Jimmy Zhong"
date: "8/16/2022"
output: html_document
---

```{r setup, include=FALSE}
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source("./init_share.R")
g(all_MAGs, low_trans, high_trans) %=% init_bins()
all_MAGs = merge_MAGs_CAZyme_peptide(all_MAGs)
MAGs_p = filter(all_MAGs, eukarya_prop < 0.05 & depth != "MIXED")

MAGs_p$CAZ_fudge = MAGs_p$percent_sect_CAZ + get.fudge(MAGs_p, "percent_sect_CAZ")
MAGs_p$pep_fudge = MAGs_p$percent_sect_pep + get.fudge(MAGs_p, "percent_sect_pep")
MAGs_p$trans_fudge = MAGs_p$percent_trans + get.fudge(MAGs_p, "percent_trans")

bi_depth = c("Shadow (SRF, DCM)", "Deep (MES, BAT)")

MAGs_p = MAGs_p %>%
  mutate(is_deep = ifelse(depth %in% c("SRF", "DCM"), bi_depth[1], bi_depth[2])) %>%
  mutate(is_deep = factor(is_deep, levels = bi_depth))
```

## Particle-associated MAGs tend to have bigger genomes

Bigger genomes is associated with higher proportion of transposases in MAGs.
Both showed in this figure (part C?), and also in the integron finder paper, and
the Touchon paper: Causes of Insertion Sequences Abundance in Prokaryotic Genomes.

The lighter color toward the top of the graph demonstrate that.

```{r}
# install.packages("viridis")
library(viridis)

left.shared = function(g) {
  g = g + ylab("complete genome size (Mbp)") +
    scale_y_log10() +
    theme_bw()
  return(g)
}

trans_scale = c(0.01, 0.03, 0.1, 0.3, 1)

id.col = c("trans_fudge", "complete_genome_size", "is_deep")

MAGs.melt = MAGs_p[c(id.col, c("CAZ_fudge", "percent_sect_pep"))] %>% 
  reshape2::melt(id = id.col) %>%
  mutate(gene = ifelse(variable == "CAZ_fudge", "CAZyme", "peptidase"))

# genome.size_particle = MAGs.melt %>%
#   ggplot(aes(x=value, y=complete_genome_size)) +
#   geom_smooth(method = "lm", se = T) +
#   scale_color_viridis(option = "H",
#                       breaks = trans_scale, labels = trans_scale,
#                       name = "Transposase abundance (%)", trans = "log") +
#   geom_point(aes(color = trans_fudge), alpha = 0.4, size = 0.7) +
#   facet_wrap(~gene, ncol = 1) +
#   xlab("secretory CAZyme/peptidase (%)\n") +
#   stat_cor(method = "spearman", cor.coef.name = "rho") +
#   scale_x_log10()

genome.size_particle = MAGs.melt %>%
  ggplot(aes(x=value, y=complete_genome_size)) +
  geom_smooth(method = "lm", se = T) +
  scale_color_viridis(option = "H",
                      breaks = trans_scale, labels = trans_scale,
                      name = "\nTransposase abundance (%)", trans = "log") +
  scale_x_log10() +
  facet_grid(is_deep~gene) +
  geom_jitter(aes(color = trans_fudge), alpha = 0.4, size = 0.7) +
  xlab("secretory CAZyme/peptidase (%)") +
  stat_cor(method = "spearman", cor.coef.name = "rho", size = 3)


MAGs.melt %>%
  ggplot(aes(x=value, y=log(trans_fudge))) +
  geom_smooth(method = "lm", se = T) +
  scale_color_viridis(option = "H",
                      breaks = trans_scale, labels = trans_scale,
                      name = "\nTransposase abundance (%)", trans = "log") +
  scale_x_log10() +
  facet_grid(is_deep~gene) +
  geom_jitter(aes(color = trans_fudge), alpha = 0.4, size = 0.7) +
  xlab("secretory CAZyme/peptidase (%)") +
  stat_cor(method = "spearman", cor.coef.name = "rho", size = 3)


genome.size_particle = left.shared(genome.size_particle) +
  theme(legend.position="top",
        legend.direction = "horizontal")  + 
  guides(color = guide_colourbar(title.position = "top",
                                 label.hjust = 0.5,
                                 barlenght = 0.8,
                                 barwidth = 10))
```

## bigger geneome, more transposases
```{r}
size_bin = cut(MAGs_p$complete_genome_size, breaks = c(0,2,3,4,5,20))
tmp = gsub("5-20", ">5",
  gsub(",", "-",
  gsub(']','', fixed=TRUE,
  gsub('(','', fixed=TRUE, size_bin))))
tmp.lvl = c("0-2","2-3","3-4","4-5",">5")

genome.size_trans.abun = MAGs_p %>%
  mutate(size_bin = factor(tmp, levels=tmp.lvl)) %>%
  ggplot(aes(x=size_bin, y=percent_trans)) + 
  geom_boxplot(outlier.alpha = 0.2) +
  stat_summary(fun.data = boxplot.give.n, geom = "text", 
               position=position_nudge(x = 0, y = 0.3)) + 
  facet_wrap(~depth, ncol= 1) +
  coord_cartesian(ylim = c(0,0.8)) +
  ylab("transposase ORF in MAGs (%)") +
  xlab("complete genome size (Mbp)") +
  theme_classic() +
  theme(legend.position = "none")
```

## deeper ocean, larger genomes 
```{r}
# install.packages("multcompView")
library("multcompView")

depth.c = c("SRF\n(n = 329)", "DCM\n(n = 339)", "MES\n(n = 479)", "BAT\n(n = 255)")

MAGs_p = MAGs_p %>%
  mutate(depth.count = ifelse(depth == "SRF", depth.c[1],
   ifelse(depth == "DCM", depth.c[2],
   ifelse(depth == "MES", depth.c[3], depth.c[4] )))) %>%
  mutate(depth.count = factor(depth.count, levels = rev(depth.c)))

aov.res = aov(complete_genome_size~depth.count, data = MAGs_p)
letters.df = data.frame(multcompLetters(TukeyHSD(aov.res)$depth.count[,4])$Letters)

colnames(letters.df)[1] = "Letter"
letters.df$depth.count = rownames(letters.df)

placement = MAGs_p %>%
  group_by(depth.count) %>%
  summarise(y.pos = quantile(complete_genome_size)[4])

letters.df = left_join(letters.df, placement) 

depth_genome.size = MAGs_p %>%
  ggplot(aes(y=complete_genome_size, x=depth.count)) +
  geom_boxplot(outlier.alpha = 0.2) +
  geom_text(data = letters.df, aes(x = depth.count, y = y.pos, label = Letter),
            size = 4, color = "black", hjust = -0.4, vjust = -0.4) +
  coord_flip()

depth_genome.size = left.shared(depth_genome.size) +
  theme(axis.title.y = element_blank())
```

## merge graph
```{r}
left = ggarrange(depth_genome.size, genome.size_particle,
                 labels = c("A", "B"),
                 ncol = 1, heights = c(2.7,7))

ggarrange(left, genome.size_trans.abun,
          labels = c("", "C"), ncol = 2, widths = c(8,5))

ggsave("out/F3_MAG_transposase_genome_size_v2.pdf",
       height = 8, width = 6)

ggsave(paste0(googleDrive.loc, "/Figure3_MAG_transposase_genome_size.pdf"),
       plot = last_plot(), height = 8, width = 7)
```

## Deep ocean MAGs are more particle-associated

Despite large differences between taxonomic groups, we will see a general
trend of high proportion of secretory CAZyme and peptidase among all CAZyme
and peptidase across different depths.

Notice that the mostly planktonic classes, Acidimicrobidae and SAR202-2,
are low in this particle-association scale. And the mostly biofilm classes,
such as Planctomycetia and Verrucomicrobia, are high in this scale.

Acidimicrobidae:
- Genomes of planktonic Acidimicrobiales: widening horizons for marine Actinobacteria by metagenomics

SAR202-2: Planktonic
- Fig 5: Particle-association lifestyle is a phylogenetically
conserved trait in bathypelagic prokaryotes

Planctomycetia:
- Planctomycetes dominate biofilms on surfaces of the kelp Laminaria hyperborea
- Planctomycetes and macroalgae, a striking association

Verrucomicrobia:
Fig 5: Particle-association lifestyle is a phylogenetically
conserved trait in bathypelagic prokaryotes

Can be both particle-associated and free-living, still have high secretory CAZyme because they are polysaccharide degrading:
Verrucomicrobia Are Candidates for Polysaccharide-Degrading Bacterioplankton in an Arctic Fjord of Svalbard: predominantly heterotrophic, carbohydrate-degrading metabolism

```{r}
MAGs_p2 = MAGs_p %>%
  group_by(Class, depth) %>%
  mutate(n = n()) %>%
  filter(n >= 5)

MAGs_p2 %>%
  ggplot(aes(x=percent_sect_CAZpep, y=fct_rev(depth))) +
  geom_boxplot(outlier.alpha = 0.3, alpha = 0.5) +
  facet_wrap(~large_classes) +
  labs(x = "secretory CAZyme and peptidase ORFs (%)",
       y = "Depth") +
  theme_classic() +
  stat_summary(fun.data = boxplot.give.nr, 
               geom = "text", position=position_nudge(x = 15, y = 0))

ggsave("redux_out/per-class-depth-secretory-percent.pdf", plot = last_plot(),
       height = 6, width = 8)
```

*The stats supporting more particle-associated MAGs in deep ocean*
```{r}
MAGs_p.stat = MAGs_p2 %>% filter(large_classes != "Others Or Unknown")
pre_CAZ = lm(percent_sect_CAZpep ~ Class, data = MAGs_p.stat)
post_CAZ = lm(percent_sect_CAZpep ~ Class + depth, data = MAGs_p.stat)
anova(pre_CAZ, post_CAZ)
summary(post_CAZ)
```

## In each Class and each depth, particle-association correlates with larger genomes

15 lower limit: The Spearman's Rank Correlation test can only be used if
there are at least 10 (ideally at least 15-15) pairs of data.

All significant correlation between %-secretory CAZyme and peptidase are positive.

```{r}
taxa_cor = MAGs_p %>%
  group_by(large_classes, depth) %>%
  mutate(n = n()) %>%
  filter(n >= 15) %>%
  summarize(p.val = cor.test(percent_sect_CAZpep, complete_genome_size, method = "spearman")$p.val,
            estimate = cor.test(percent_sect_CAZpep, complete_genome_size, method = "spearman")$estimate,
            count = n()) %>%
  adjust_pvalue(method = "BH") %>%
  add_significance()

sign_sort = rbind(taxa_cor[taxa_cor$p.val.adj.signif != "ns", ],
                  taxa_cor[taxa_cor$p.val.adj.signif == "ns", ]) 

sign_sort = sign_sort %>%
  # mutate(p_text = round(p.val.adj, 4)) %>%
  select(c("large_classes","depth","estimate","count","p.val.adj","p.val.adj.signif"))

colnames(sign_sort) = c("class", "depth", "correlation estimate", "count", "adjusted p-value", "significance")
write.csv(sign_sort,"redux_out/per-class-depth-sect-gsize-corr.csv", row.names = FALSE)
```


