---
title: "Graph everything with metagenomic particle association as x"
author: "Jimmy Zhong"
date: "8/17/2022"
output: html_document
---

```{r setup, include=FALSE}
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source("./init_share.R")
g(DNA_tara, RNA_tara, DNA_RNA_tara) %=% init_tara()
mala_cov = init_mala_cov()
```

## Worry about scaling later
```{r}
defense_percent_scale <- c(0.5, 1, 2)
defense_scale <- log10(defense_percent_scale/100)

CAZ_percent_scale <- c(0.8, 0.4, 0.2, 0.1)
CAZ_scale <- log2(CAZ_percent_scale/100)
```

## More initiation
```{r}
mala_cov$size_fraction <- ifelse(mala_cov$lower_filter_size == "0.2", "planktonic", "particle")
DNA_tara$size_fraction <- ifelse(DNA_tara$upper_size_dna == "1.6", "planktonic", "particle")

sel = colnames(mala_cov)[grep("log_dna|percent", colnames(mala_cov))]
sel = c(sel, "Ocean_DNA", "Layer_DNA", "size_fraction")

bd_to_graph <- rbind(mala_cov[sel], DNA_tara[sel])
bd_to_graph$Layer_DNA <- factor(bd_to_graph$Layer_DNA, levels = c("SRF","DCM","MES","BAT"))
```


## particle association and defense mechanism ORFs
```{r}
bd_to_graph %>%
  filter(Layer_DNA != "MIX") %>%
  ggplot(aes(x=percent_dna_sect_CAZpep, y=log_dna_defense)) +
  # facet_wrap(~Layer_DNA, ncol = 2) + # scales = "free"
  # scale_y_continuous(breaks = defense_scale, labels = defense_percent_scale, limits=c(-2.3, -1.65)) + 
  # scale_x_continuous(breaks = CAZ_scale, labels = CAZ_percent_scale) +
  geom_point(aes(color=Layer_DNA)) +
  theme_classic() +
  ylab("% DNA reads mapped to defense mechanism ORFs") +
  # xlab("Average percentage of secretory CAZyme and peptidase (DNA)") +
  geom_smooth(method = "lm", se = T)  +
  theme()
```


## particle association and signal transduction ORFs
```{r}
bd_to_graph %>%
  filter(Layer_DNA != "MIX") %>%
  ggplot(aes(x=percent_dna_sect_CAZpep, y=log_dna_signalT)) +
  # facet_wrap(~Layer_DNA, ncol = 2) + # scales = "free"
  geom_point(aes(color=Layer_DNA)) +
  theme_classic() +
  geom_smooth(method = "lm", se = T)  +
  theme()
```


```{r}
bd_to_graph %>%
  filter(Layer_DNA != "MIX") %>%
  ggplot(aes(x=percent_dna_sect_CAZpep, y=log_dna_replication)) +
  # facet_wrap(~Layer_DNA, ncol = 2) + # scales = "free"
  geom_point(aes(color=Layer_DNA)) +
  theme_classic() +
  geom_smooth(method = "lm", se = T)  +
  theme()
```



DNA_RNA_tara %>%
  filter(Layer_RNA != "MIX") %>%
  filter(Ocean %in% c("Indian Ocean", "North Atlantic", "North Pacific", "Arctic Ocean",
                      "South Atlantic", "South Pacific")) %>%
  ggplot(aes(x=log(avg_percent.y), y=log_dna_defense)) +
  scale_color_brewer(palette="Set1")+
  facet_wrap(~Layer_RNA, ncol = 2) +
  geom_point(aes(color = Ocean)) + 
  theme_classic() +
  # ylab("% DNA reads mapped to transposase") +
  xlab("Average percentage of secretory CAZyme and peptidase (RNA)") +
  geom_smooth(method = "lm", se = F)


bd_to_graph %>%
  filter(!Layer_DNA %in% c("MIX",NA)) %>%
  ggplot(aes(x=percent_sect_CAZ, y=log_dna_defense)) +
  facet_wrap(~Layer_DNA, ncol = 2) + # scales = "free"
  scale_y_continuous(breaks = defense_scale, labels = defense_percent_scale) + # , limits=c(-2.3, -1.65) 
  geom_point(aes(color=size_fraction)) +
  theme_classic() +
  ylab("% DNA reads mapped to defense mechanism ORFs") +
  # xlab("abundance of secretory CAZyme in a metagenome") +
  xlab("secretory CAZenzyme among all CAZyme (%)") +
  geom_smooth(method = "lm", se = F)  +
  theme()

bd_to_graph %>%
  ggplot(aes(x=DNA_sect_pep, y=percent_sect_pep))+
  geom_point(aes(color=size_fraction, shape = Layer_DNA)) +
  xlab("abundance of secretory peptidase in a metagenome") +
  ylab("abundance of secretory peptidase in all peptidase")

bd_to_graph %>%
  ggplot(aes(x=log_dna_sect_CAZ, y=percent_sect_CAZ))+
  geom_point(aes(color=size_fraction, shape = Layer_DNA)) +
  xlab("abundance of secretory CAZyme in a metagenome") +
  ylab("abundance of secretory CAZyme in all CAZyme")

plot(DNA_sect_pep ~ percent_sect_pep, bd_to_graph)

# used in gen_figure5_integron_finder_category_v2.R
p_bd <- bd_to_graph %>%
  filter(Layer_DNA != "MIX") %>%
  ggplot(aes(x=log_dna_biofilm, y=log_dna_defense)) +
  facet_wrap(~Layer_DNA, ncol = 1) + # scales = "free"
  scale_y_continuous(breaks = defense_scale, labels = defense_percent_scale, limits=c(-2.3, -1.75)) + 
  scale_x_continuous(breaks = biofilm_scale2, labels = biofilm_percent_scale2, limits=c(-4, -3.25)) + # outlier
  geom_point() + # aes(color = size_fraction)
  theme_classic() +
  ylab("% DNA reads mapped to defense mechanism ORFs") +
  xlab("% DNA reads mapped to\nbiofilm ORFs") +
  geom_smooth(method = "lm", se = F)  +
  theme() # legend.position = "none"

defense_biofilm.lm <- lm(log_dna_defense~Layer_DNA + log_dna_biofilm, bd_to_graph)
summary(lm(log_dna_defense~Layer_DNA, bd_to_graph))
summary(defense_biofilm.lm)


