---
title: "Graph everything with metagenomic particle association as x"
author: "Jimmy Zhong"
date: "8/17/2022"
output: html_document
---

```{r setup, include=FALSE}
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source("./init_share.R")
g(DNA_tara, RNA_tara, DNA_RNA_tara) %=% init_tara()
mala_cov = init_mala_cov()
mala_cov$size_fraction <- ifelse(mala_cov$lower_filter_size == "0.2", "planktonic", "particle")
DNA_tara$size_fraction <- ifelse(DNA_tara$upper_size_dna == "1.6", "planktonic", "particle")
```

## Worry about scaling later
```{r}
defense_percent_scale <- c(0.5, 1, 2)
defense_scale <- log10(defense_percent_scale/100)

CAZ_percent_scale <- c(0.8, 0.4, 0.2, 0.1)
CAZ_scale <- log2(CAZ_percent_scale/100)
```

## More initiation
```{r}
sel = colnames(mala_cov)[grep("log_dna", colnames(mala_cov))]
sel = sel[!grepl("CAZ|pep|log_dna_trans", sel)]
independent_var = c("percent_dna_sect_CAZpep","Layer_DNA","Ocean_DNA")
sel2 = c(sel, independent_var)

bd_to_graph <- rbind(mala_cov[sel2], DNA_tara[sel2])
bd_to_graph$Layer_DNA <- factor(bd_to_graph$Layer_DNA, levels = c("SRF","DCM","MES","BAT"))

bd_melt = bd_to_graph %>%
  filter(Ocean_DNA != "Arctic") %>%
  filter(!is.na(Layer_DNA)) %>%
  reshape2::melt(id.var = independent_var) %>%
  mutate(value = as.numeric(value))

colnames(bd_melt) = c(independent_var, "COG", "COG_abun")

# see each correlation
bd_melt %>%
  ggplot(aes(x = percent_dna_sect_CAZpep, y = COG_abun)) +
  facet_wrap(~COG, scale = "free_y") +
  geom_point(aes(color = Layer_DNA)) +
  geom_smooth(method = "lm", se = T)
```

## statistics on correlations
```{r}
sel.cor = colnames(DNA_tara)[grep("DNA", colnames(DNA_tara))]
sel.cor2 = c(sel.cor[11:32], independent_var)

COG.corr <- rbind(mala_cov[sel.cor2], DNA_tara[sel.cor2]) #  %>%
  filter(Ocean_DNA != "Arctic")
COG.corr$Layer_DNA <- factor(COG.corr$Layer_DNA, levels = c("SRF","DCM","MES","BAT"))

COG.melt = COG.corr %>%
  reshape2::melt(id.var = independent_var)
colnames(COG.melt) = c(independent_var,"COG","COG_abun")
```


## Lienar regresssion based correlation looks good!
```{r}
base_info = COG.corr[independent_var]
out = c()
for (COG in sel.cor2[grepl("DNA_", sel.cor2)]) {
  print(COG)
  y_info = cbind(COG.corr[c(COG, "percent_dna_sect_CAZpep","Layer_DNA")])
  y_info['COG_log'] = log2(COG.corr[COG])
  depth.lm = lm(formula=COG_log ~ Layer_DNA, data = y_info)
  depth.CAZpep.lm = lm(formula=COG_log ~ Layer_DNA + percent_dna_sect_CAZpep, data = y_info)
  p.val = anova(depth.lm, depth.CAZpep.lm)$`Pr(>F)`[2]
  out = rbind(out, c(COG, p.val, 
                     depth.CAZpep.lm$coefficients['percent_dna_sect_CAZpep']))
}

out = as.data.frame(out)
colnames(out) = c("COG", "p.val", "coefficient")
out$p.val = as.numeric(out$p.val)

out2 = out %>%
  adjust_pvalue(method = "BH") %>%
  arrange(p.val.adj) %>%
  add_significance()

out_display = out2 %>%
  mutate(COG_category = gsub("DNA_", "", out2$COG),
         signif = p.val.adj.signif) %>%
  select(COG_category, p.val.adj, coefficient, signif) %>%
  mutate(coefficient = round(as.numeric(coefficient), 4))
```

> out_display
                           COG_category    p.val.adj coefficient signif
1        Signal_transduction_mechanisms 4.839823e-07      0.0478   ****
2                         Cell_motility 3.027711e-03      0.0357     **
3                    Defense_mechanisms 1.319531e-02      0.0258      *
4                             Cell_wall 1.167937e-01      0.0192     ns
5                         Transcription 1.167937e-01      0.0191     ns

> out_display
                           COG_category    p.val.adj coefficient signif
1        Signal_transduction_mechanisms 1.539041e-15      0.0476   ****
2                         Cell_motility 1.121542e-08      0.0414   ****
3                    Defense_mechanisms 1.108282e-04      0.0227    ***
4        Mobilome_prophages_transposons 1.108282e-04      0.0502    ***
5                  Extracellular_struct 2.900663e-04      0.0335    ***
6                         Transcription 2.900663e-04      0.0220    ***
7                          Cytoskeleton 3.938139e-03      0.0308     **
8                             Cell_wall 1.719129e-02      0.0147      *
9             Intracellular_trafficking 9.665419e-02      0.0115     ns
10       Posttranslational_modification 1.075020e-01      0.0111     ns


## Plotting the COG categories that have correlations across depths
```{r}
col_COG = c("DNA_Signal_transduction_mechanisms",
"DNA_Defense_mechanisms","DNA_Cell_motility")

sel3 = c(col_COG, independent_var)
COG_viz = rbind(mala_cov[sel3], DNA_tara[sel3]) %>%
  filter(Ocean_DNA != "Arctic")

viz_melt = COG_viz %>%
  filter(Layer_DNA != "MIX") %>%
  reshape2::melt(id.var = independent_var) %>%
  mutate(COG_abun = as.numeric(value)) %>%
  # mutate(log_abun = log2(COG_abun) ) %>%
  select(c(independent_var,"variable","log_abun","COG_abun"))

# install.packages("ggh4x")
library("ggh4x")

viz_melt$COG = viz_melt$variable
pretty_COG=c("Signal Transduction", "Defense Mechanism", "Cell motility")
for (i in 1:length(col_COG)){
  viz_melt$COG = gsub(col_COG[i],pretty_COG[i], viz_melt$COG)
}

# see each correlation
viz_melt %>%
  mutate(Layer_DNA = factor(Layer_DNA,
    levels=c("SRF","DCM","MES","BAT"))) %>%
  ggplot(aes(x = percent_dna_sect_CAZpep,
             y = COG_abun * 100)) +
  facet_grid2(Layer_DNA~COG,
              scales = "free_y",
              independent = "y") + 
  scale_y_log10()+
  geom_point(aes(color = Ocean_DNA)) + # aes(color = Ocean_DNA)
  geom_smooth(method = "lm", se = T) + 
  labs(color = "Ocean",
       y = "Gene abundance of COG category (%)",
       x = "Secretory CAZyme and peptidase (%)") +
  theme_bw()

ggsave(filename = "redux_out/particle-COG-association-DNA.pdf",
       last_plot(), width = 7, height = 6)
```


## Exploring from this point, not acutally used

## Spearman correlation

There are too many depths, a little too noisy
```{r}
COG.stats = COG.melt %>%
  filter(Layer_DNA != "MIX") %>%
  group_by(COG, Layer_DNA) %>%
  summarize(p.val = cor.test(percent_dna_sect_CAZpep, 
                             COG_abun, method = "spearman")$p.val,
            estimate = cor.test(percent_dna_sect_CAZpep,
                             COG_abun, method = "spearman")$estimate,
            count = n()) %>%
  adjust_pvalue(method = "BH")

COG.stats$p.adj = p.adjust(COG.stats$p.val) 

COG.stats = COG.stats %>%
  add_significance() %>%
  filter(p.val.adj < 0.05)
```

## Plotting the COG categories that have correlations across depths
```{r}
RNA_id = c("percent_rna_sect_CAZpep","Layer_RNA","Ocean")
sel3_RNA = c("RNA_Signal_transduction_mechanisms",
"RNA_Defense_mechanisms","RNA_Cell_motility")

RNA_viz = RNA_tara[c(RNA_id, sel3_RNA)] %>%
  filter(Layer_RNA != "MIX") %>%
  filter(Ocean != "Arctic")

RNA_viz_melt = RNA_viz %>%
  reshape2::melt(id.var = RNA_id) %>%
  mutate(log_abun = as.numeric(value)) %>%
  select(c(RNA_id,"variable","log_abun"))

# library("ggh4x")

RNA_viz_melt$COG = RNA_viz_melt$variable
pretty_COG=c("Signal Transduction", "Defense Mechanism", "Cell motility")
for (i in 1:length(sel3_RNA)){
  RNA_viz_melt$COG = gsub(sel3_RNA[i], pretty_COG[i], RNA_viz_melt$COG)
}

# see each correlation
RNA_viz_melt %>%
  mutate(Layer_RNA = factor(Layer_RNA,
    levels=c("SRF","DCM","MES","BAT"))) %>%
  ggplot(aes(x = percent_rna_sect_CAZpep,
             y = log_abun * 1000)) +
  facet_grid2(Layer_RNA~COG,
              scales = "free_y",
              independent = "y") + 
  scale_y_log10()+
  geom_point(aes(color = Ocean)) + # aes(color = Ocean_DNA)
  geom_smooth(method = "lm", se = T) +
  theme_bw()

rowSums(DNA_tara[sel.cor[11:32]])
rowSums(mala_cov[sel.cor[11:32]])
```

