---
  title: "Graph metagenomic expression ratio"
author: "Jimmy Zhong"
date: "8/21/2022"
output: html_document
---
  
## Initialization
  
```{r}
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source("./init_share.R")
# g(DNA_tara, RNA_tara, DNA_RNA_tara, depth_comparison) %=% init_tara()
mala_cov <- init_mala_cov()
g(DNA_Metadata, RNA_Metadata) %=% init_tara_md(factorize = FALSE)

size.frac = c("0.22 - 1.6 (*Tara*)", "0.22 - 3 (*Tara*)",
              "0.2 - 0.8 (Malaspina)", "0.8 - 5 (Malaspina)",
              "0.22 - 1.6 (*Tara*) Arctic", "0.22 - 3 (*Tara*) Arctic")
```

```{r}
sum_reads_into_row = function(hit_table_f, base_path){
  sum_reads_per_file = function(singel_hit_table, base_path){
    sample_name = str_replace(singel_hit_table, paste(base_path,"/",sep = ""), "")
    sample_name = str_replace(sample_name, "_all_hits_table.txt", "")
    if (file.info(singel_hit_table)$size < 60) { # no reads mapped
      print(paste("empty table:",singel_hit_table))
      return(c(sample_name, 0))
    } else {
      ht = suppressMessages(read_delim(singel_hit_table, delim = "\t",
                                       escape_double = FALSE, trim_ws = TRUE))
      return(c(sample_name, sum(ht$Read_counts)))
    }
    return(df)
  }
  df = mapply(sum_reads_per_file, hit_table_f, base_path)
  df = t(as.data.frame(df))
  rownames(df) <- 1:nrow(df)
  colnames(df) = c("sample","euk_read_count")
  df = as.data.frame(df)
  df = transform(df, euk_read_count = as.numeric(euk_read_count))
  return(df)
}

hit_table = "*_all_hits_table.txt"
base02="../identify_eukaryotes/EukDetect-deep/02_size_fraction/filtering"
base08="../identify_eukaryotes/EukDetect-deep/08_size_fraction/filtering"
deep_frac02 = Sys.glob(file.path(base02, hit_table))
deep_frac08 = Sys.glob(file.path(base08, hit_table))

euk_read_count02 = sum_reads_into_row(deep_frac02, base02)
euk_read_count02$filter_size = size.frac[3]
euk_read_count08 = sum_reads_into_row(deep_frac08, base08)
euk_read_count08$filter_size = size.frac[4]
deep_euk_count = rbind(euk_read_count02, euk_read_count08)
deep_euk_count$depth = "BAT"

# base_big_SRF = "../identify_eukaryotes/EukDetect-SRF_5-20/filtering"
# base_big_DCM = "../identify_eukaryotes/EukDetect-DCM_5-20/filtering"
# big_size_SRF_f = Sys.glob(file.path(base_big_SRF, hit_table))
# big_size_DCM_f = Sys.glob(file.path(base_big_DCM, hit_table))
# 
# big_size_SRF = sum_reads_into_row(big_size_SRF_f, base_big_SRF)
# big_size_SRF$depth = "SRF"
# big_size_DCM = sum_reads_into_row(big_size_DCM_f, base_big_DCM)
# big_size_DCM$depth = "DCM"
# big_size_count = rbind(big_size_SRF, big_size_DCM)
# big_size_count$filter_size <- "size5-20"

base_tara="../identify_eukaryotes/EukDetect-normal/filtering"
tara_normal = Sys.glob(file.path(base_tara, hit_table))
euk_tara = sum_reads_into_row(tara_normal, base_tara)
euk_tara$ocean = "normal"

# base_arct="../identify_eukaryotes/EukDetect-arctic/filtering"
# tara_arctic = Sys.glob(file.path(base_arct, hit_table))
# euk_arctic = sum_reads_into_row(tara_arctic, base_arct)

# euk_arctic$ocean = "arctic"
# euk_tara = rbind(euk_tara, euk_arctic)

euk_tara$euk_read_count = as.numeric(euk_tara$euk_read_count)
```


```{r}
sel_col = c("Layer_DNA","upper_size_dna")
match_md_row = function(x){
  md_row = DNA_Metadata[DNA_Metadata$ENA_Run_ID %like% x["sample"], sel_col]
  if (nrow(md_row) > 0) {
    df = c(x, unlist(md_row))
    return(df)
  } else {
    print(paste("cannot find", x["sample"]))
  }
}

tmp = as.data.frame(apply(euk_tara, 1, match_md_row))
tmp = as.data.frame(t(tmp))
tmp = tmp[!grepl("cannot find", tmp$sample), ]

merged_tara = tmp %>%
  mutate(depth = Layer_DNA,
         filter_size = ifelse(upper_size_dna == "1.6", 
                       ifelse(ocean == "arctic", size.frac[5], size.frac[1]),
                       ifelse(ocean == "arctic", size.frac[6], size.frac[2])),
         euk_read_count = as.numeric(as.character(euk_read_count)))

sel_col2 = c("sample", "filter_size", "euk_read_count", "depth")
depth_cat = rbind(merged_tara[sel_col2], deep_euk_count[sel_col2]) # big_size_count[sel_col2]
total_reads = read_csv("../identify_eukaryotes/count-ocean.out")
all_cat = merge(depth_cat, total_reads, by = "sample")

all_cat$euk_prop = all_cat$euk_read_count/all_cat$num_reads

all_cat = all_cat %>% 
  mutate(depth = factor(depth, levels = c("BAT", "MES", "DCM", "SRF")),
         filter_size = factor(filter_size, levels = size.frac),
         euk_prop = euk_prop + get.fudge(all_cat, "euk_prop"))
```

## Viz
```{r}
all_cat %>%
  filter(depth != "MIX") %>%
  ggplot(aes(x = euk_prop * 100, y = depth, fill = filter_size)) +
  geom_boxplot(outlier.alpha = 0.5) +
  # geom_point(shape = 21, alpha = 0.5,
  #            position = position_jitterdodge(jitter.width = 0.3)) +
  labs(x = "reads mapped to eukaryotic markers (%)",
       fill = expression(paste("Size Fraction (", mu, "m)"))) +
  theme_bw() +
  scale_x_log10() +
  theme(legend.text = ggtext::element_markdown(),
        axis.text.x = element_text(color="black"),
        axis.text.y = element_text(color="black"))

ggsave("out/S7-eukaryote-proportions.pdf",
       height = 3, width = 6)
```

