---
title: "Graph everything with metagenomic particle association as x"
author: "Jimmy Zhong"
date: "8/17/2022"
output: html_document
---

```{r setup, include=FALSE}
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source("./init_share.R")
g(DNA_tara, RNA_tara, DNA_RNA_tara) %=% init_tara()
mala_cov = init_mala_cov()
mala_cov$size_fraction <- ifelse(mala_cov$lower_filter_size == "0.2", "planktonic", "particle")
DNA_tara$size_fraction <- ifelse(DNA_tara$upper_size_dna == "1.6", "planktonic", "particle")

DNA_tara$sample = DNA_tara$ENA_Run_ID
```

## More initiation
```{r}
sel = colnames(mala_cov)[grep("log_dna", colnames(mala_cov))]
sel = sel[!grepl("CAZ|pep|log_dna_trans", sel)]
independent_var = c("percent_dna_sect_CAZpep","Layer_DNA",
  "percent_dna_sect_pep", "percent_dna_sect_CAZ",
  "sample","Ocean_DNA","DNA_biofilm")
sel2 = c(sel, independent_var)

sel.cor = colnames(DNA_tara)[grep("DNA", colnames(DNA_tara))]
all_COGs = sel.cor[12:33]
sel.cor2 = c(all_COGs, independent_var)
COG.corr = rbind(mala_cov[sel.cor2], DNA_tara[sel.cor2])
```

## Make corr.melt
```{r}
d_level = c("SRF","DCM","MES","BAT")
COG.corr = COG.corr %>% 
  mutate(Layer_DNA = factor(Layer_DNA, levels = d_level)) %>%
  filter(Ocean_DNA != "Arctic")

COG.melt = COG.corr %>%
  reshape2::melt(id.var = independent_var)
colnames(COG.melt) = c(independent_var,"COG","COG_abun")
```


## Spearman correlation
```{r}
get_COG_corr = function(COG.melt, feature) {
  tmp = COG.melt
  tmp['fea'] = COG.melt[feature]

  tmp = tmp %>%
    filter(!is.na(Layer_DNA))
    # filter(COG != "DNA_Chromatin_structure_and_dynamics")
  
  COG.stats = tmp %>%
  group_by(COG) %>%
  summarize(p.val = cor.test(fea, COG_abun, method = "spearman")$p.val,
            estimate = cor.test(fea, COG_abun, method = "sp")$estimate,
            count = n()) %>%
  adjust_pvalue(method = "BY") %>%
  add_significance()
  return(COG.stats)
}

COG.stats1 = get_COG_corr(COG.melt, "percent_dna_sect_pep")
COG.stats2 = get_COG_corr(COG.melt, "percent_dna_sect_CAZ")

COG.stats1$Depth = "peptidase"
COG.stats2$Depth = "CAZyme"
COG.stats.comb = rbind(COG.stats1, COG.stats2)

# COG.stats.pep = COG.stats1 %>%
#   arrange(COG, is_deep) %>% #,Layer_DNA
#   mutate(Depth = paste0("pep-", is_deep))
#   # paste0("pep-", Layer_DNA)
# 
# COG.stats.CAZ = COG.stats2 %>%
#   arrange(COG) %>% #,Layer_DNA
#   mutate(Depth = paste0("CAZ-", is_deep))
# 
# COG.stats.comb = rbind(COG.stats.CAZ, COG.stats.pep)
# COG.stats.comb$Depth = factor(COG.stats.comb$Depth,
#     levels = c("CAZ-BAT","CAZ-tara", "pep-BAT", "pep-tara"))

# sig_COG = unique(COG.stats.comb[COG.stats.comb$p.val.adj < 0.05, ]$COG)

sig_COG = unique(COG.stats.comb$COG)
```

## Doing the heatmap

Some colnames preparation
```{r}
col_to_index = function(df, index_col){
  return(df %>% remove_rownames %>% column_to_rownames(var=index_col))
}

cast_long_list = function(corr_long_list, var1, var2, value) {
  estimate_2d = reshape2::dcast(corr_long_list,
                                paste(var1,var2,sep="~"),value.var=value)
  rownames_tmp = unlist(estimate_2d[var1])
  estimate_2d = col_to_index(estimate_2d, var1)
  if (value == "estimate") {
    estimate_2d = estimate_2d %>% apply(2, as.numeric) 
  }
  rownames(estimate_2d) = rownames_tmp
  return(estimate_2d)
}

COG.graph = COG.stats.comb[COG.stats.comb$COG %in% sig_COG, ]
COG.graph$est.sig = ifelse(COG.graph$p.val.adj < 0.05, 
                           round(COG.graph$estimate, 3), "")

COG.corr.2D = cast_long_list(COG.graph, "COG", "Depth", "estimate")

old_names = rownames(COG.corr.2D)
pretty = gsub("TM", "transport and metabolism", 
         gsub("_", " ", 
         gsub("DNA_", "", old_names)))

pretty = gsub("Cell wall", "Cell wall/membrane/envelope biogenesis", pretty)
pretty = gsub("lular struct", "lular structures", pretty)
pretty = gsub("Translation ribosomal structure", "Translation, ribosomal structure", pretty)

rownames(COG.corr.2D) = pretty

for (i in 1:length(old_names)){
  COG.graph$COG = gsub(old_names[i], pretty[i], COG.graph$COG)  
}
```

Get the clustering
```{r}
library("ggdendro")
COG_dendro = as.dendrogram(hclust(d = dist(x = COG.corr.2D)))
dendro_graph = ggdendrogram(data = COG_dendro, rotate = FALSE) + 
  coord_flip() +
  scale_x_continuous(expand = c(0,0)) +
  scale_y_reverse() +
  theme_dendro() +
  theme(plot.margin = margin(t = 5, b = 5, r = 0, l = 5))

blank_plot = ggplot() + theme_void() +
  theme(plot.margin = margin(t = 5, b = 5, r = 0, l = 5))
integron_order = rownames(COG.corr.2D)[order.dendrogram(COG_dendro)]
```

Doing the heatmap
```{r}
COG_ggheatmap = COG.graph %>%
  mutate(COG = factor(COG, levels = integron_order)) %>%
  ggplot(aes(x=Depth, y=COG)) + 
  geom_tile(aes(fill=estimate)) +
  scale_fill_gradient2(low="blue", mid="white", high="red") +
  geom_text(aes(label=p.val.adj.signif), color="black", size=3) +
  theme_classic() +
  labs(fill = "COG correlation with\nsecretory CAZyme & \npeptidase genes (%)") +
  scale_y_discrete(position = "right") + 
  theme (axis.text.x = element_text(angle = -35, vjust = 0.5, size = 10),
         axis.title.x = element_blank(), # ggtext::element_markdown(),
         legend.background = element_rect(fill="transparent",
                                          color="transparent"),
         axis.title.y = element_blank(),
         axis.text.y = element_blank(),
         axis.line.y = element_blank(),
         legend.title=element_text(size=10),
         plot.margin = margin(t = 5, b = 5, r = 0, l = 0),
         legend.position = c(6.25, 0.8)) 
```


Combining the heatmap with the graph *p_cassette_category* from
*redux_fig5_integron_finder_category.R*
```{r}
left = ggarrange(blank_plot, dendro_graph, blank_plot, 
                  heights = c(0.02, 0.8, 0.07), ncol = 1)

ggarrange(left, COG_ggheatmap, p_cassette_category, 
          ncol = 3, widths=c(0.5,0.8,5)) # , labels = c("A", "B")

ggsave("out/link-pure-COG-prop-with-integron.pdf",
       width = 8, height = 5.5)

ggsave(paste0(googleDrive.loc, "/Figure5_link-COG-abun-with-integron.pdf"),
       width = 8, height = 5.5)
```

# Supplementary scatter plot: see each correlation
```{r}
simplify = function(vec){
  vec = gsub("transport and metabolism", "T.M.", vec)
  vec = gsub("structure", "struct", vec)
  vec = gsub(" prophages transposons", "", vec)
  vec = gsub("wall/membrane/envelope biogenesis", "wall", vec)
  vec = gsub("transduction mechanisms", "transduction", vec)
  vec = gsub("Replication recombination and repair",
             "Replication", vec)
  vec = gsub("Translation, ribosomal struct", "Translation", vec)
  vec = gsub("Energy production and conversion", "Energy", vec)
  return(vec)
}

pt.COG.graph = COG.melt %>% 
  filter(!is.na(Layer_DNA))

for (i in 1:length(old_names)){
  pt.COG.graph$COG = gsub(old_names[i], pretty[i], pt.COG.graph$COG)  
}

integron_order_sim = rev(simplify(integron_order))

pt.COG.graph = pt.COG.graph %>%
  filter(COG %in% integron_order) %>%
  mutate(is_deep = ifelse(Layer_DNA == "BAT", "BAT", "no"),
    COG2 = factor(simplify(COG), levels = integron_order_sim),
    COG_abun = COG_abun * 100,
    COG_percent_abun = COG_abun/COG_sum_coverage * 100)

# no significantly different between integron and normal
# metagenomes
out = c("Intracellular trafficking", "Secondary metabolites",
        "cell cycle control")

gen_COG_graph = function(fea1, fea2){
  pt.COG.graph %>%
  filter(!COG2 %in% out) %>%
  ggplot(aes_string(x = fea1, y = fea2,
             color = "is_deep")) +
  facet_wrap(~COG2, scale = "free_y", ncol = 4) +
  geom_point(aes(shape = Layer_DNA), size = 1) +
  geom_smooth(method = "lm", se = F) +
  scale_y_sqrt() +
  labs(shape = "Depth",
       color = "Is BAT") +
  theme_classic()
}

gen_COG_graph("percent_dna_sect_CAZ", "COG_abun") +
  ylab("COG category abundance (%)") + 
  xlab("Secretory CAZyme (%)")

# ggsave("out/particle-COG-association-CAZ.pdf",
ggsave("out/particle-COG-raw-prop-association-CAZ.pdf",
       width = 9.5, height = 9)

gen_COG_graph("percent_dna_sect_pep", "COG_abun") +
  ylab("COG category abundance (%)") + 
  xlab("Secretory peptidase (%)")

# ggsave("out/particle-COG-association-pep.pdf",
ggsave("out/particle-COG-raw-prop-association-pep.pdf",
       width = 9.5, height = 9)
```



## Exploring from this point, not acutally used

```{r}
library(vegan)

in_pcoa = COG.corr[all_COGs]
# in_pcoa = in_pcoa/(c(mala_COG_sum, tara_COG_sum))
rownames(in_pcoa) = COG.corr$sample
md_pcoa = COG.corr[independent_var]
rownames(md_pcoa) = COG.corr$sample

COG.bray = vegdist(in_pcoa)
COG.b.pcoa = cmdscale(COG.bray, k = (nrow(COG.corr[all_COGs]) - 1), 
                      eig = TRUE)
pcoa_p = data.frame(PC1 = COG.b.pcoa$points[,1],
                    PC2 = COG.b.pcoa$points[,2])

pcoa_p = merge(pcoa_p, md_pcoa, by=0) #work-in metadata

#viz
pcoa_p %>%
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_point(aes(color = Layer_DNA)) +
  theme_classic() +
  labs(color = "Depth")
  
ggsave("out/PCoA_COGs_pure_prop_depth.pdf",
       width = 4.5, height = 4)
```




## Can't really do linear regression

The normality of residual assumption does not meet.

```{r}
out = c()
# I don't have Chromatin_structure in my integrons, don't care about it.
# [all_COGs != "DNA_Chromatin_structure_and_dynamics"]
for (COG in all_COGs) {
  print(COG)
  y_info = cbind(COG.corr[c(COG, independent_var)])
  # y_info['COG_log'] = log2(COG.corr[COG])
  y_info['COG_log'] = log(COG.corr[COG]/COG.corr$COG_sum_coverage)
  depth.lm = lm(formula=COG_log ~ Layer_DNA + Ocean_DNA, data = y_info)
  depth.CAZpep.lm = lm(formula=COG_log ~ Layer_DNA + Ocean_DNA + log(percent_dna_sect_CAZpep), data = y_info)
  print(shapiro.test(depth.CAZpep.lm$residuals))
  p.val = anova(depth.lm, depth.CAZpep.lm)$`Pr(>F)`[2]
  out = rbind(out, c(COG, p.val,
        depth.CAZpep.lm$coefficients['log(percent_dna_sect_CAZpep)']))
}

out = as.data.frame(out)
colnames(out) = c("COG", "p.val", "coefficient")

# I can't really assume that different COGs are independent,
# so I can't use BH p.value adjust
out$p.val = as.numeric(out$p.val)
out$p.val.adj = p.adjust(out$p.val)

out2 = out %>%
  arrange(coefficient) %>%
  add_significance(output.col = "signif")

out_display = out2 %>%
  mutate(COG_category = gsub("DNA_", "", out2$COG)) %>%
  filter(p.val.adj < 0.05) %>%
  select(COG_category, p.val.adj, coefficient, signif) %>%
  mutate(coefficient = round(as.numeric(coefficient), 4))

# write.csv(out_display,"redux_out/particle-COG-corr.csv", row.names = FALSE)
```

> out_display
                           COG_category    p.val.adj coefficient signif
1        Signal_transduction_mechanisms 1.539041e-15      0.0476   ****
2                         Cell_motility 2.141125e-08      0.0414   ****
3                    Defense_mechanisms 3.967492e-04      0.0227    ***
4        Mobilome_prophages_transposons 3.967492e-04      0.0502    ***
5                         Transcription 1.231117e-03      0.0220     **
6                  Extracellular_struct 1.344853e-03      0.0335     **
7                          Cytoskeleton 2.004871e-02      0.0308      *
8                             Cell_wall 9.377067e-02      0.0147     ns

## Plotting the COG categories that have correlations across depths
```{r}
col_COG = c("DNA_Signal_transduction_mechanisms", "DNA_Cell_motility",
"DNA_Defense_mechanisms", "DNA_Mobilome_prophages_transposons",
"DNA_Transcription", "DNA_Extracellular_struct", "DNA_Cytoskeleton")

sel3 = c(col_COG, independent_var)
COG_viz = rbind(mala_cov[sel3], DNA_tara[sel3])

viz_melt = COG_viz %>%
  filter(Layer_DNA != "MIX") %>%
  reshape2::melt(id.var = independent_var) %>%
  mutate(COG_abun = as.numeric(value)) %>%
  mutate(log_abun = log2(COG_abun) ) %>%
  select(c(independent_var,"variable","log_abun","COG_abun")) %>%
  mutate(Layer_DNA = factor(Layer_DNA,
    levels=c("SRF","DCM","MES","BAT")))

# install.packages("ggh4x")
# library("ggh4x")

viz_melt$COG = viz_melt$variable
pretty_COG=c("Signal Transduction", "Cell motility", "Defense Mechanism", 
             "Mobilome", "Transcription","Extracellular structures",
             "Cytoskeleton")
for (i in 1:length(col_COG)){
  viz_melt$COG = gsub(col_COG[i],pretty_COG[i], viz_melt$COG)
}
viz_melt$COG = factor(viz_melt$COG, levels = pretty_COG)
```

# see each correlation
```{r}
viz_melt %>%
  ggplot(aes(x = percent_dna_sect_CAZpep,
             y = COG_abun * 100)) +
  # facet_grid2(Layer_DNA~COG,
  #             scales = "free_y",
  #             independent = "y") + 
  facet_wrap(~COG, scales = "free_y", ncol = 4)+
  scale_y_log10()+
  geom_point(aes(color = Layer_DNA), size = 1) + # aes(color = Ocean_DNA)
  # geom_smooth(method = "lm", se = T) + 
  labs(color = "Depth",
       y = "Gene abundance of COG category (%)",
       x = "Secretory CAZyme and peptidase (%)") +
  theme_classic() +
  theme(legend.position = c(0.87, 0.25))

# ggsave(filename = "redux_out/particle-COG-association.pdf",
#        last_plot(), width = 8.5, height = 4.5)
```

## Biofilm special
```{r}
melt.biofilm = merge(viz_melt, sum.COG, by = "sample") 

melt.biofilm = melt.biofilm %>%
  mutate(percent_COG = COG_abun/COG_sum_coverage)

melt.biofilm %>%
  ggplot(aes(x = DNA_biofilm * 10000,
             y = percent_COG)) +
  facet_wrap(~COG, scales = "free_y", ncol = 4)+
  # scale_y_log10()+
  scale_x_log10()+
  geom_point(aes(color = Layer_DNA), size = 1) + # aes(color = Ocean_DNA)
  labs(color = "Depth",
       y = "Gene abundance of COG category (%)",
       x = "Biofilm gene abundance (% * 100)") +
  theme_classic() +
  theme(legend.position = c(0.87, 0.25))

# ggsave(filename = "redux_out/particle-COG-association.pdf",
#        last_plot(), width = 8.5, height = 4.5)
```

## Stats on biofilm
```{r}
out.biofilm = c()
for (COG in all_COGs) {
  print(COG)
  y_info = cbind(COG.corr[c(COG, independent_var)])
  y_info['COG_log'] = log2(COG.corr[COG])
  y_info['log_biofilm'] = log2(COG.corr['DNA_biofilm'])
  depth.lm = lm(COG_log ~ Layer_DNA, data = y_info)
  biofilm.lm = lm(COG_log ~ Layer_DNA + log_biofilm, data = y_info)
  p.val = anova(depth.lm, biofilm.lm)$`Pr(>F)`[2]
  out.biofilm = rbind(out.biofilm, c(COG, p.val,
                      biofilm.lm$coefficients['log_biofilm']))
}

out.biofilm = as.data.frame(out.biofilm)
colnames(out.biofilm) = c("COG", "p.val", "coefficient")

out$p.val = as.numeric(out$p.val)
out$p.val.adj = p.adjust(out$p.val)

out2 = out %>%
  arrange(p.val.adj) %>%
  add_significance()

out_display = out2 %>%
  mutate(COG_category = gsub("DNA_", "", out2$COG),
         signif = p.val.adj.signif) %>%
  select(COG_category, p.val.adj, coefficient, signif) %>%
  mutate(coefficient = round(as.numeric(coefficient), 4))
```

## Plotting the COG categories that have correlations across depths
```{r}
RNA_id = c("percent_rna_sect_CAZpep","Layer_RNA","Ocean")
sel3_RNA = c("RNA_Signal_transduction_mechanisms",
"RNA_Defense_mechanisms","RNA_Cell_motility")

RNA_viz = RNA_tara[c(RNA_id, sel3_RNA)] %>%
  filter(Layer_RNA != "MIX") %>%
  filter(Ocean != "Arctic")

RNA_viz_melt = RNA_viz %>%
  reshape2::melt(id.var = RNA_id) %>%
  mutate(log_abun = as.numeric(value)) %>%
  select(c(RNA_id,"variable","log_abun"))

# library("ggh4x")

RNA_viz_melt$COG = RNA_viz_melt$variable
pretty_COG=c("Signal Transduction", "Defense Mechanism", "Cell motility")
for (i in 1:length(sel3_RNA)){
  RNA_viz_melt$COG = gsub(sel3_RNA[i], pretty_COG[i], RNA_viz_melt$COG)
}

# see each correlation
RNA_viz_melt %>%
  mutate(Layer_RNA = factor(Layer_RNA,
    levels=c("SRF","DCM","MES","BAT"))) %>%
  ggplot(aes(x = percent_rna_sect_CAZpep,
             y = log_abun * 1000)) +
  facet_grid2(Layer_RNA~COG,
              scales = "free_y",
              independent = "y") + 
  scale_y_log10()+
  geom_point(aes(color = Ocean)) + # aes(color = Ocean_DNA)
  geom_smooth(method = "lm", se = T) +
  theme_bw()

rowSums(DNA_tara[sel.cor[11:32]])
rowSums(mala_cov[sel.cor[11:32]])




COG_heatmap = pheatmap(mat = COG.corr.2D,
         display_numbers = COG.stars,
         cluster_cols = F,
         cluster_rows = T,
         show_rownames = T,
         show_colnames = T,
         legend = T,
         # fontsize = ,
         # cellwidth = 6.5, cellheight = 6.5,
         scale = "none")

integron_order = rownames(COG.corr.2D[COG_heatmap$tree_row[["order"]],])
gg_COG_heatmap = ggplotify::as.ggplot(COG_heatmap)
```

