---
title: "the COG category of the cassette and non-cassette metagenomic ORFs"
author: "Jimmy Zhong"
date: "8/17/2022"
output: html_document
---

```{r setup, include=FALSE}
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source("./init_share.R")
summary = init_integron_category()
```

total 1225 cassette "proteins" from Malaspina, 7294 from Tara Oceans metagenomes.
359 cassette from Malaspina receive COG function calls, compared to 1951 from Tara.

Malaspina total ORFs: 2438506, with COG annotation: 1657370 -> annotation rate: 0.67967
South Atlatnic total: 8263808, COG annota: 4405933 -> rate: 0.5331
North Atlantic total: 4718433, COG annota: 2759276 -> rate: 0.5847
Mediterrean total: 3818905, COG: 2270527 -> rate: 0.5945
Red sea total: 1988319, COG: 1145209 -> rate: 0.57597

source, total_cassettes, functional_cassettes
total            25178   9929 -> 39.4%
ARS               2417   840
CPC               2194   747
EAC               1680   677
IN                 687   271
MED                907   442
NAT               1679   861
NP                2701   1051
RS                 509   267
SAT               4205   1757
SP                5174   1110
deep              3025   1206 -> 40.0%

COG_integrons[COG_integrons.function == "nan"] # 16177 rows, 16177/25178 = 0.6425054
COG_integrons['function'] = COG_integrons.function.str.replace(pattern, "Function unknown", regex=True)
COG_integrons[COG_integrons.function == "Function unknown"] # 17439 rows, 17439/25178 = 0.69262848518

## Find differences between cassettes and normal metagenomic

Use hypergeometric test because this is drawing without replacement

See:
https://stats.stackexchange.com/questions/369006/statistical-test-for-comparing-two-frequencies-with-r

```{r}
# options(scipen = 0)
integron_total = sum(summary$integron_count)
norm_total = sum(summary$normal_count)

COG_enrichment.stat = t(apply(summary, 1, function(r){
  group1 = integron_total
  int_count = as.numeric(r["integron_count"])
  nor_count = as.numeric(r["normal_count"])
  group2 = int_count + nor_count
  p.val1 = phyper(int_count, nor_count, norm_total-nor_count, integron_total, lower.tail= FALSE)
  # test for over-representation (enrichment)
  p.val2 = phyper(int_count-1, nor_count, norm_total-nor_count, integron_total, lower.tail= TRUE)
  return(c(r['COG_function'], p.val1, p.val2))
}))

enrich = COG_enrichment.stat %>%
  as.data.frame() %>%
  reshape2::melt(id.var = "COG_function") %>%
  select(c("COG_function", "value"))

colnames(enrich) = c("COG","p.val")
enrich = enrich %>%
  mutate(p.val = as.numeric(p.val)) %>%
  adjust_pvalue(method = "BY") %>%
  add_significance(cutpoints = c(0, 0.05, 1),
                   symbols = c("*","")) %>%
  filter(p.val < 0.5)
```

## Needs redux_fig5_biofilm_defense_corr
```{r}
rownames(enrich) = enrich$COG
COG_sig = enrich[integron_order, ]$p.val.adj.signif
integron_order2 = paste(integron_order, COG_sig, sep = "")

graphing = summary[summary$COG_function %in% integron_order, ]
rownames(graphing) = graphing$COG_function
graphing = graphing[integron_order, ]
graphing$COG = paste(graphing$COG_function, COG_sig, sep = "")
graphing = graphing[c("COG", "integron_prop", "normal_prop")]
g.melt = graphing %>% reshape2::melt(id.vars=c("COG"))
g.melt$type = factor(g.melt$variable,levels = c("normal_prop", "integron_prop"))

cas_legend = sprintf("Cassette ORFs (n = %s)", as.character(integron_total))
# nor_legend <- expression(paste("all other ORFs in", italic("Tara"), " Oceans"))
nor_legend = "all ORFs in metagenomes"
```


## REQUIRE redux_fig5_biofilm_defense_corr.Rmd
```{r}
p_cassette_category = g.melt %>% 
  mutate(COG = factor(COG,levels=integron_order2))%>%
  ggplot(aes(fill=type, y=value, x=COG)) +
  geom_bar(position="dodge", stat="identity", width = 0.66) + 
  coord_flip(ylim = c(0, 0.15), clip = "off") +
  ylab("Proprotion to total ORF calls\n\n") + 
  theme_classic() +
  guides(fill = guide_legend(reverse = TRUE, title = "COG gene-calls for:")) +
  scale_fill_manual(values=c("dark gray", "orange"), 
                    labels = c(nor_legend, cas_legend)) +
  theme(axis.text.y = element_text(hjust = 0.5), # size=6, 
    axis.title.y = element_blank(),
    plot.background = element_rect(fill='transparent', color=NA),
    legend.background = element_rect(fill="transparent", color="transparent")) 
# ggsave(filename = "redux_out/integron-COG-category-only.pdf",
#        last_plot(), width = 5.2, height = 6)
```

## COG enrichement
> enrich$COG_function
 [1] "Replication, recombination and repair"  "Defense mechanisms"                    
 [3] "Mobilome: prophages, transposons"       "Transcription"                         
 [5] "Signal transduction mechanisms"         "Cell motility"                         
 [7] "Extracellular structures"               "Cytoskeleton"                          
 [9] "Translation, ribosomal structure..."    "Amino acid TM*"                        
[11] "Cell wall/membrane/envelope biogenesis" "Posttranslational modification..."     
[13] "Energy production and conversion"       "Coenzyme TM"                          
[15] "Lipid TM"                              "Carbohydrate TM"                      
[17] "Inorganic ion TM*"                      "Nucleotide TM"

## p_bd sees gen_figure5_biofilm_defense_corr.R
```{r}

  filter(COG_function %in% integron_order) %>%
  mutate(COG = COG_function + COG_sig)
  filter(integron_prop > 0.001) %>%
  arrange(integron_prop) %>%
  mutate(integron_val_rank = integron_prop) %>%

ggarrange(p_cassette_category, p_bd, labels = c("A", "B"), 
          ncol = 2, nrow = 1, widths = c(0.7, 0.35))

ggsave("F5_cassette-COG-cateogory_biofilm-vs-defense.png", 
       plot = last_plot(),
       height = 8,
       width = 8)
ggsave("F5_cassette-COG-cateogory_biofilm-vs-defense.pdf", 
       plot = last_plot(),
       height = 8,
       width = 8)
```
