---
title: "Particle associated lifestyle and genome size and depths of MAGs"
author: "Jimmy Zhong"
date: "8/16/2022"
output: html_document
---

```{r setup, include=FALSE}
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source("./init_share.R")
all_MAGs = init_bins()
all_MAGs = merge_MAGs_CAZyme_peptide(all_MAGs)
MAGs_p = filter(all_MAGs, eukarya_prop < 0.05 & depth != "MIXED")

MAGs_p$CAZ_fudge = MAGs_p$percent_sect_CAZ + get.fudge(MAGs_p, "percent_sect_CAZ")
MAGs_p$pep_fudge = MAGs_p$percent_sect_pep + get.fudge(MAGs_p, "percent_sect_pep")
MAGs_p$trans_fudge = MAGs_p$percent_trans + get.fudge(MAGs_p, "percent_trans")

bi_depth = c("Shallow (SRF, DCM)", "Deep (MES, BAT)")

MAGs_p = MAGs_p %>%
  mutate(is_deep = ifelse(depth %in% c("SRF", "DCM"), bi_depth[1], bi_depth[2])) %>%
  mutate(is_deep = factor(is_deep, levels = bi_depth))
```

## Particle-associated MAGs tend to have bigger genomes

Bigger genomes is associated with higher proportion of transposases in MAGs.
Both showed in this figure (part C?), and also in the integron finder paper, and
the Touchon paper: Causes of Insertion Sequences Abundance in Prokaryotic Genomes.

The lighter color toward the top of the graph demonstrate that.

```{r}
# install.packages("viridis")
library(viridis)

left.shared = function(g) {
  g = g + ylab("complete genome size (Mbp)") +
    scale_y_log10() +
    theme_bw() +
    theme(axis.text.x = element_text(color="black"),
          axis.text.y = element_text(color="black"))
  return(g)
}

trans_scale = c(0.01, 0.03, 0.1, 0.3, 1)

id.col = c("trans_fudge", "complete_genome_size", "is_deep")
```

## Cutting lifestyles
```{r}
size_CAZ = cut(MAGs_p$CAZ_fudge, breaks = c(-1, 10, 20, 100))
size_pep = cut(MAGs_p$percent_sect_pep, breaks = c(-1, 20, 40, 100))

x.gsub = function(vec) {return (gsub(",", "-",
  gsub(']','', fixed=TRUE, gsub('-1', '0', 
  gsub('(','', fixed=TRUE, vec))))) }

size_CAZ = x.gsub(size_CAZ)
size_pep = x.gsub(size_pep)

MAGs_p = MAGs_p %>% 
  mutate(particle.CAZ = size_CAZ,
         particle.pep = size_pep)
```

## Per Taxa analysis
```{r}
graph_taxon = c("Flavobacteria", "Verrucomicrobia", "Gammaproteobacteria")
g.tax = c("Flavobacteria", "Verrucomicrobia", "\u03b3-proteobacteria")
ct = c("high", "normal", "low")

col_list = c("depth", "percent_trans", "trans_fudge",
             "percent_sect_CAZ", "percent_sect_pep", "Class")

g_tax = MAGs_p[col_list] %>% 
  filter(Class %in% graph_taxon) %>%
  mutate(class_trans = ifelse(Class == g.tax[1], ct[1],
                       ifelse(Class == g.tax[2], ct[2], ct[3])),
         Class = ifelse(Class == graph_taxon[3], g.tax[3], Class),
         Class = factor(Class, levels = g.tax))

color_code <- rev(c("cyan","steelblue","yellow"))

tax_depth <- g_tax %>%
  ggplot(aes(x=percent_trans, y = fct_rev(depth), fill = class_trans)) +
  scale_fill_manual(values = color_code)+
  facet_wrap(~Class, ncol = 3) +
  geom_boxplot(outlier.alpha = 0.3,
               outlier.shape = 21,
               outlier.color = NA) + 
  ylab("depth")+
  scale_x_continuous(breaks = c(0, 0.2, 0.4, 0.6), limits=c(0, 0.8)) +
  stat_summary(fun.data = boxplot.give.n, 
               geom = "text", 
               position=position_nudge(x = 0.4, y = 0)) + 
  theme_classic() +
  theme(legend.position = "none",
        axis.title.x=element_blank(),
        axis.title.y=element_blank())

ls_order <- c("planktonic", "mixed", "particle", "all")

g_tax_label <- c("High %-transposase classes (\u0251- \u03b2- \u03b3- \u03b4- proteobact.)", 
                "Normal transposase abundance classes", 
                "Low %-transposase classes (Flavobact., Acidimicrobidae, etc.)")
g_tax_col <- color_code

# lab_text <- expression("Taxonomical "*italic("Class")*" of")

g_tax %>% # tax_lifestyle <- 
  ggplot(aes(x = trans_fudge, y = percent_sect_CAZ)) +
  facet_wrap(~Class, ncol = 3) +
  geom_point(aes(color = depth)) +
  scale_x_log10() + 
  geom_smooth(method = "lm") +
  stat_cor(method = "spearman", cor.coef.name = "rho") + 
  labs(# y = "secretory peptidase (%)",
       y = "secretory CAZyme (%)",
       x = "transposase ORFs in MAGs (%)") +
  theme_classic()
  theme(# strip.text.x = element_blank(),
        axis.title.y = element_blank(),
        legend.title = element_blank(),
        legend.justification='left',
        legend.margin = margin(0),
        legend.position="bottom",
        legend.direction="vertical")

```


```{r}
gen.gs.particle = function(x.lab.text, sig.pos, cut.gene){
  tmp = MAGs_p
  tmp['gene'] = cut.gene
  
  gs.particle.stats = tmp %>%
    group_by(depth) %>%
    wilcox_test(complete_genome_size~gene) %>%
    adjust_pvalue(method = "BY") %>%
    filter(p.adj < 0.05) %>%
    add_significance() %>%
    add_xy_position()
  
  print(gs.particle.stats)
  
  # gs.particle.stats$y.position = sig.pos
  
  tmp %>%
    ggplot(aes(x=gene, y=complete_genome_size)) +
    geom_boxplot(outlier.alpha = 0.1, varwidth = TRUE) +
    facet_wrap(~depth, scale = "free_x", ncol = 1, strip.position="right") +
    stat_pvalue_manual(gs.particle.stats, tip.length = 0.01,
       label = "p.adj.signif", hide.ns = TRUE ) + 
    labs(y = "complete genome size (Mbp)", x = x.lab.text) +
    coord_cartesian(ylim = c(0.7, 8)) +
    theme_classic() +
    theme(axis.text.x = element_text(color="black"),
          axis.text.y = element_text(color="black")) +
    stat_summary(fun.data = boxplot.give.nr, 
      geom = "text", position=position_nudge(x = 0, y = 0.3))
}

CAZ.gz = gen.gs.particle("secretory CAZyme (%)", 
                         c(6.5, 7, 7.2, 7.7), size_CAZ)
pep.gz = gen.gs.particle("secretory peptidase (%)", 
                         c(6.5, 7.1, 7.8, 7.35, 8.05, 7.65), size_pep)
```

## bigger geneome, more transposases
```{r}
size_bin = cut(MAGs_p$complete_genome_size, breaks = c(0,2,3,4,5,20))
tmp = gsub("5-20", ">5",
  gsub(",", "-",
  gsub(']','', fixed=TRUE,
  gsub('(','', fixed=TRUE, size_bin))))
tmp.lvl = c("0-2","2-3","3-4","4-5",">5")

genome.size_trans.abun = MAGs_p %>%
  mutate(size_bin = factor(tmp, levels=tmp.lvl)) %>%
  ggplot(aes(x=size_bin, y=percent_trans)) + 
  geom_boxplot(outlier.alpha = 0.2) +
  stat_summary(fun.data = boxplot.give.n, geom = "text", 
               position=position_nudge(x = 0, y = 0.3)) + 
  facet_wrap(~depth, ncol= 1) +
  coord_cartesian(ylim = c(0,0.8)) +
  ylab("transposase ORF in MAGs (%)") +
  xlab("complete genome size (Mbp)") +
  theme_classic() +
  theme(legend.position = "none",
        axis.text.x = element_text(color="black"),
        axis.text.y = element_text(color="black"))
```

## deeper ocean, larger genomes 
```{r}
depth.c = c("SRF\n(n = 329)", "DCM\n(n = 339)", "MES\n(n = 479)", "BAT\n(n = 255)")

MAGs_p = MAGs_p %>%
  mutate(depth.count = ifelse(depth == "SRF", depth.c[1],
   ifelse(depth == "DCM", depth.c[2],
   ifelse(depth == "MES", depth.c[3], depth.c[4] )))) %>%
  mutate(depth.count = factor(depth.count, levels = rev(depth.c)))

aov.res = aov(complete_genome_size~depth.count, data = MAGs_p)
letters.df = data.frame(multcompView::multcompLetters(TukeyHSD(aov.res)$depth.count[,4])$Letters)
colnames(letters.df)[1] = "Letter"
letters.df['depth.count'] = rownames(letters.df)
placement = MAGs_p %>%
  group_by(depth.count) %>%
  summarise(y.pos = quantile(complete_genome_size)[4])

letters.df = left_join(letters.df, placement)

depth_genome.size = MAGs_p %>%
  ggplot(aes(y=complete_genome_size, x=depth.count)) +
  geom_boxplot(outlier.alpha = 0.2) +
  geom_text(data = letters.df, 
            aes(x = depth.count, y = y.pos, label = Letter),
            size = 4, color = "black", hjust = -0.4, vjust = -0.4) +
  coord_flip()

depth_genome.size = left.shared(depth_genome.size) +
  theme(axis.title.y = element_blank())
```

## merge graph
```{r}
left = ggarrange(depth_genome.size, genome.size_particle,
                 labels = c("A", "B"),
                 ncol = 1, heights = c(2.3,7))

ggarrange(left, genome.size_trans.abun,
          labels = c("", "C"), ncol = 2, widths = c(8, 5))

ggsave("out/F3_MAG_transposase_genome_size_v2.pdf",
       height = 8, width = 7)

ggsave(paste0(googleDrive.loc, "/Figure3_MAG_transposase_genome_size.pdf"),
       plot = last_plot(), height = 8, width = 7)
```

## Deep ocean MAGs are more particle-associated

Despite large differences between taxonomic groups, we will see a general
trend of high proportion of secretory CAZyme and peptidase among all CAZyme
and peptidase across different depths.

Notice that the mostly planktonic classes, Acidimicrobidae and SAR202-2,
are low in this particle-association scale. And the mostly biofilm classes,
such as Planctomycetia and Verrucomicrobia, are high in this scale.

Acidimicrobidae:
- Genomes of planktonic Acidimicrobiales: widening horizons for marine Actinobacteria by metagenomics

Dehalococcoides:
Characterization of the community structure of a dechlorinating mixed culture and comparisons of gene expression in planktonic and biofloc-associated "Dehalococcoides" and Methanospirillum species.

SAR202-2: Planktonic
- Fig 5: Particle-association lifestyle is a phylogenetically
conserved trait in bathypelagic prokaryotes

Planctomycetia:
- Planctomycetes dominate biofilms on surfaces of the kelp Laminaria hyperborea
- Planctomycetes and macroalgae, a striking association

Verrucomicrobia:
Fig 5: Particle-association lifestyle is a phylogenetically
conserved trait in bathypelagic prokaryotes

Sphingobacterium:
Metagenomic analysis of size-fractionated picoplankton in a marine oxygen minimum zone


Can be both particle-associated and free-living, still have high secretory CAZyme because they are polysaccharide degrading:
Verrucomicrobia Are Candidates for Polysaccharide-Degrading Bacterioplankton in an Arctic Fjord of Svalbard: predominantly heterotrophic, carbohydrate-degrading metabolism

# Class particle association
```{r}
class.median = MAGs_p %>%
  filter(!is.na(percent_sect_pep)) %>%
  group_by(Class) %>%
  summarise(count =n(),
    median_CAZ = mean(percent_sect_CAZ),
    median_pep = mean(percent_sect_pep)) %>%
  filter(count >= 10 & !Class == "Others Or Unknown" & !Class == "other") %>%
  arrange(median_CAZ + median_pep)
```

```{r}
class.count = paste0(class.median$Class, " (n=", class.median$count, ")")

MAGs_p2 = MAGs_p %>% filter(Class %in% class.median$Class)

for (c in 1:length(class.median$Class)) {
  MAGs_p2$Class = gsub(class.median$Class[c], class.count[c], MAGs_p2$Class)
}

MAGs_p2 = MAGs_p2 %>%
  mutate(Class = factor(Class, levels = class.count),
         CAZyme = percent_sect_CAZ + get.fudge(MAGs_p2, "percent_sect_CAZ"),
         peptidase = percent_sect_pep + get.fudge(MAGs_p2, "percent_sect_pep"))

id.v2 = c("complete_genome_size", "is_deep", "Class")
tax.melt = MAGs_p2[c(id.v2, c("CAZyme", "peptidase"))] %>% #"trans_fudge"
  reshape2::melt(id = id.v2) %>%
  mutate(gene = variable,
         percent.CAZ.pep = value)

tax.melt %>%
  ggplot(aes(x=percent.CAZ.pep, y=Class)) +
  geom_boxplot(outlier.alpha = 0.3, alpha = 0.5) +
  facet_wrap(~gene, scale = "free_x") +
  labs(x = "secretory CAZyme/peptidase ORFs (%)") +
  theme_classic() +
  theme(axis.title.y = element_blank(),
        axis.text.x = element_text(color="black"),
        axis.text.y = element_text(color="black"))

ggsave("out/S5-class-secretory-CAZyme-peptidase.pdf",
       height = 5, width = 6)
```


*The stats supporting more particle-associated MAGs in deep ocean*
```{r}
MAGs_p.stat = MAGs_p2 %>% filter(large_classes != "Others Or Unknown")
pre_CAZ = lm(percent_sect_CAZpep ~ Class, data = MAGs_p.stat)
post_CAZ = lm(percent_sect_CAZpep ~ Class + depth, data = MAGs_p.stat)
anova(pre_CAZ, post_CAZ)
summary(post_CAZ)
```

## Individual Class analysis (boxplot)
```{r}
MAGs_p3 = MAGs_p %>% filter(Class %in% class.median$Class)

id.v3 = c("complete_genome_size", "depth", "percent_trans", "particle.CAZ", "particle.pep", "Class")
tax.melt3 = MAGs_p3[id.v3] %>%
  group_by(Class, depth) %>%
  mutate(class.dpeth.count = n()) %>%
  group_by(Class, particle.CAZ) %>%
  mutate(class.CAZ.count = n()) %>%
  group_by(Class, particle.pep) %>%
  mutate(class.pep.count = n(),
         trans_fudge = percent_trans + get.fudge(MAGs_p3, "percent_trans"))

more.than.1.group = function(df, var1, var2){
  all.var1 = unlist(unique(df[var1]))
  great.than.1 = c()
  for (e in 1:length(all.var1)) {
    tmp = df[df[var1] == all.var1[e], ]
    all.var2 =unlist(unique(tmp[var2]))
    if (length(all.var2) > 1) {
      great.than.1 = c(great.than.1, all.var1[e])
    }
  }
  return(great.than.1)
}

taxa.share = function(g) {
  g + 
    facet_wrap(~Class, ncol = 4) +
    geom_boxplot(outlier.alpha = 0.5,
                 outlier.shape = 21) +
    scale_x_log10() +
    xlab("transposase ORFs in MAGs (%)") + 
    # scale_x_continuous(breaks = c(0, 0.2, 0.4, 0.6), limits=c(0, 0.8)) +
    stat_summary(fun.data = boxplot.give.n, 
                 geom = "text", 
                 position=position_nudge(x = 0.4, y = 0)) + 
    theme_classic() +
    theme(legend.position = "none",
          axis.text.x = element_text(color="black"),
          axis.text.y = element_text(color="black"))
}

more.than.1.depth = tax.melt3 %>%
  filter(class.dpeth.count > 5) %>%
  more.than.1.group("Class", "depth")

more.than.1.CAZ = tax.melt3 %>%
  filter(class.CAZ.count > 5) %>%
  more.than.1.group("Class", "particle.CAZ")

more.than.1.pep = tax.melt3 %>%
  filter(class.pep.count > 5) %>%
  more.than.1.group("Class", "particle.pep")


depth.trans = tax.melt3 %>%
  filter(class.dpeth.count > 5 & Class %in% more.than.1.depth) %>%
  ggplot(aes(x=trans_fudge, y = fct_rev(depth))) + # , fill = class_trans
  ylab("depth")

CAZyme.trans = tax.melt3 %>%
  filter(class.CAZ.count > 5 & Class %in% more.than.1.CAZ) %>%
  ggplot(aes(x=trans_fudge, y = fct_rev(particle.CAZ))) + # , fill = class_trans
  ylab("secretory CAZyme (%)")

pep.trans = tax.melt3 %>%
  filter(class.pep.count > 5 & Class %in% more.than.1.pep) %>%
  ggplot(aes(x=trans_fudge, y = fct_rev(particle.pep))) + # , fill = class_trans
  ylab("secretory peptidase (%)")

taxa.share(depht.trans)
taxa.share(CAZyme.trans)
taxa.share(pep.trans)
```

## Individual Class analysis (numerical scattorplot)
```{r}
MAGs_p3 = MAGs_p %>% filter(Class %in% class.median$Class)

big.MAGs = MAGs_p3[c("complete_genome_size", "depth", "trans_fudge", "percent_sect_CAZ", "percent_sect_pep", "Class")]

big.MAGs %>% # CAZyme.trans = 
  ggplot(aes(x=trans_fudge, y = percent_sect_CAZ)) + # , fill = class_trans
  ylab("secretory CAZyme (%)") +
  facet_wrap(~Class, ncol = 4) +
  geom_point(aes(color = depth)) +
  scale_x_log10() +
  xlab("transposase ORFs in MAGs (%)") +
  stat_cor(method = "spearman", cor.coef.name = "rho")

pep.trans = tax.melt3 %>%
  filter(class.pep.count > 5 & Class %in% more.than.1.pep) %>%
  ggplot(aes(x=trans_fudge, y = fct_rev(particle.pep))) + # , fill = class_trans
  ylab("secretory peptidase (%)")

taxa.share(depht.trans)
taxa.share(CAZyme.trans)
taxa.share(pep.trans)
```

## In each Class and each depth, particle-association correlates with larger genomes

15 lower limit: The Spearman's Rank Correlation test can only be used if
there are at least 10 (ideally at least 15-15) pairs of data.

All significant correlation between %-secretory CAZyme and peptidase are positive.

```{r}
taxa_cor = MAGs_p %>%
  group_by(large_classes, depth) %>%
  mutate(n = n()) %>%
  filter(n >= 15) %>%
  summarize(p.val = cor.test(percent_sect_CAZpep, complete_genome_size, method = "spearman")$p.val,
            estimate = cor.test(percent_sect_CAZpep, complete_genome_size, method = "spearman")$estimate,
            count = n()) %>%
  adjust_pvalue(method = "BH") %>%
  add_significance()

sign_sort = rbind(taxa_cor[taxa_cor$p.val.adj.signif != "ns", ],
                  taxa_cor[taxa_cor$p.val.adj.signif == "ns", ]) 

sign_sort = sign_sort %>%
  # mutate(p_text = round(p.val.adj, 4)) %>%
  select(c("large_classes","depth","estimate","count","p.val.adj","p.val.adj.signif"))

colnames(sign_sort) = c("class", "depth", "correlation estimate", "count", "adjusted p-value", "significance")
write.csv(sign_sort,"redux_out/per-class-depth-sect-gsize-corr.csv", row.names = FALSE)
```

## genome size and transposase abundance via cut
```{r}
MAGs_p %>%
  ggplot(aes(x=complete_genome_size, y=percent_trans)) + 
  geom_point(alpha = 0.3, size = 1) +
  stat_cor(method = "spearman", cor.coef.name = "rho",
           label.x.npc = 0,
           label.y.npc = 0.5, 
           aes(label = paste(..r.label.., ..p.., sep = "~"))) +
    # aes(label = paste(..r.label.., cut(..p.., 
    # breaks = c(-Inf, 0.0001, 0.001, 0.01, 0.05, Inf),
    # labels = c("'****'", "'***'", "'**'", "'*'", "'ns'")),
    # sep = "~"))) +
  facet_wrap(~depth, ncol= 1) +
  coord_cartesian(ylim = c(0,0.8)) +
  ylab("transposase ORF in MAGs (%)") +
  xlab("complete genome size (Mbp)") +
  theme_classic() +
  geom_smooth(se = T, method = "lm", formula = y ~ x + I(x^2)) +
  # scale_x_continuous(limits = c(1, 15), trans = "log10") +
  # scale_x_log10() +
  theme(legend.position = "none")
```


## linear regression between genome size and secretory CAZyme
```{r}
MAGs.melt = MAGs_p[c(id.col, c("CAZ_fudge", "percent_sect_pep"))] %>% 
  reshape2::melt(id = id.col) %>%
  mutate(gene = ifelse(variable == "CAZ_fudge", "CAZyme", "peptidase"))

genome.size_particle = MAGs.melt %>%
  ggplot(aes(x=value, y=complete_genome_size)) +
  geom_smooth(method = "lm", se = T) +
  scale_color_viridis(option = "H",
                      breaks = trans_scale, labels = trans_scale,
                      name = "\nTransposase abundance (%)", trans = "log") +
  scale_x_log10() +
  facet_grid(is_deep~gene) +
  geom_jitter(aes(color = trans_fudge), alpha = 0.4, size = 0.7) +
  xlab("secretory CAZyme/peptidase (%)") +
  stat_cor(method = "spearman", cor.coef.name = "rho", size = 3) +
  theme(axis.text.x = element_text(color="black"),
        axis.text.y = element_text(color="black"))

genome.size_particle = left.shared(genome.size_particle) +
  theme(legend.position="top",
        legend.direction = "horizontal",
        axis.text.x = element_text(color="black"),
        axis.text.y = element_text(color="black")) + 
  guides(color = guide_colourbar(title.position = "top",
                                 label.hjust = 0.5,
                                 barlenght = 0.8,
                                 barwidth = 10))

```