---
title: "Graph metagenomic expression ratio"
author: "Jimmy Zhong"
date: "8/21/2022"
output: html_document
---

## Initialization
```{r}
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source("./init_share.R")
g(DNA_tara, RNA_tara, DNA_RNA_tara) %=% init_tara()
# options(scipen=10000)

DNA_RNA = DNA_RNA_tara %>%
  filter(Layer_RNA %in% c("SRF", "DCM", "MES")) %>%
  get_expression()
```

```{r}
dna_cols = colnames(DNA_RNA)[grepl("DNA_", colnames(DNA_RNA))]
COG.dna_cols = dna_cols[7:28]
DNA.COG.sum = rowSums(DNA_RNA[COG.dna_cols])

COG.rna_cols = gsub("DNA", "RNA", COG.dna_cols)
RNA.COG.sum = rowSums(DNA_RNA[COG.rna_cols])

check.df = data.frame(DNA_RNA[c('connector_DNA','Ocean_DNA', "Layer_RNA")], DNA.COG.sum, RNA.COG.sum)
check.df$ratio = check.df$RNA.COG.sum / check.df$DNA.COG.sum

ggplot(check.df, aes(y = Ocean_DNA, x = ratio)) +
  geom_boxplot() +
  geom_point()

boxplot(ratio~Layer_RNA, check.df)
```


## Looking at expression rate

## Run this for integron-enriched COGs
```{r}
gene_col =c("Extracellular_struct", "Mobilome_prophages_transposons",
            "Defense_mechanisms", "Replication_recombination_and_repair",
            "Signal_transduction_mechanisms", "Transcription",
            "Cell_motility", "transposase")

pretty.exp = c("Extracellular structures", "Mobilome...",
  "Defense mechanisms", "Replication...", "Signal transduction",
  "Transcription", "Cell motility", "Transposase")

out.f = "/Figure6-integron-COGs-expression-ratios.pdf"
height = 3
sig.bar.pos = log10(c(5, 7,
                     2.5, 3.4,
                     1.1, 1.3,
                     1.1, 1.5,
                     2.2, 3,
                     3, 4,
                     7,10,17))
```


## Run this for others COGs
```{r}
gene_col = c("Lipid_TM", "Coenzyme_TM",
             "Energy_production_and_conversion", "Cell_wall",
             "Amino_acid_TM", "Cytoskeleton", "Inorganic_ion_TM",
             "Nucleotide_TM", "Chromatin_structure_and_dynamics",
             "Cell_cycle_control", "Intracellular_trafficking",
             "Translation_ribosomal_structure", "Carbohydrate_TM", 
             "Posttranslational_modification", "Secondary_metabolites")

pretty.exp = c("Lipid T.M.", "Coenzyme T.M.",
               "Energy production...", "Cell wall",
                "Amino acid T.M.", "Cytoskeleton", "Inorganic ion T.M.",
               "Nucleotide T.M.", "Chromatin structure...",
               "Cell cycle control", "Intracellular trafficking",
               "Translation ribosomal...", "Carbohydrate T.M.", 
               "Posttranslational...", "Secondary metabolites")

out.f = "/Figure6-other-COGs-expression-ratios.pdf"
height = 6
sig.bar.pos = log10(c(1.2, 1.6,
                     1.1,
                     1.1, 1.5,
                     3, 4, 6.5,
                     3, 4,
                     2.5, 3.5, 6,
                     3, 4,
                     5, 6,
                     2.7,
                     2.5, 3.5,
                     1.3))
```


```{r}
exp_cols = paste0(gene_col, "_exp")

DNA_RNA$depth = factor(DNA_RNA$Layer_RNA, levels = c("MES","DCM","SRF"))
base.cols = c("depth", "connector_DNA", "percent_dna_sect_CAZ", "percent_dna_sect_pep")
target = DNA_RNA[c(base.cols, exp_cols)]
exp.melt = target %>% reshape2::melt(id.var = base.cols)
colnames(exp.melt) = c(base.col, "gene", "expression")

old.gene = unique(exp.melt$gene)
exp.melt$pretty.gene = exp.melt$gene
for (g in 1:length(old.gene)) {
  exp.melt$pretty.gene = gsub(old.gene[g], pretty.exp[g], exp.melt$pretty.gene)
}

exp.melt$pretty.gene = factor(exp.melt$pretty.gene, levels = pretty.exp)
```


# actual graphing
```{r}
exp.stats = exp.melt %>%
  group_by(pretty.gene) %>%
  t_test(expression~depth) %>%
  # wilcox_test(expression~depth) %>%
  adjust_pvalue(method = "BH") %>%
  filter(p.adj < 0.05) %>%
  add_significance() %>% 
  add_xy_position()

exp.stats$y.position = sig.bar.pos

exp.melt %>%
  # filter(expression < 12) %>% # an extreme outlier for transposase, SRFs
  ggplot(aes(y = expression, x = depth)) +
  facet_wrap(~pretty.gene, scales = "free_x", ncol = 4) + # 
  scale_y_continuous(trans = "log10",
    labels = function(x) ifelse(x == 0, "0", x)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.12, width = 0.2) +
  stat_pvalue_manual(exp.stats, tip.length = 0.01,
    label = "p.adj.signif", hide.ns = TRUE ) +
  coord_flip() +
  theme_classic() +
  ylab("RNA/DNA ratio (transcript / gene abundance(%) per sample)") +
  theme(axis.title.y = element_blank())

ggsave(paste0("out", out.f), height = height, width = 7)
ggsave(paste0(googleDrive.loc, out.f), width = 7, height = height)
```


## direct correlation between partcle association and genes, not too useful
```{r}
exp.melt %>%
  filter(expression < 12) %>% # an extreme outlier for transposase, SRFs
  ggplot(aes(y = expression, x = percent_dna_sect_CAZ)) +
  facet_wrap(~gene) +
  # scale_y_continuous(trans = "log10", labels = function(x) ifelse(x == 0, "0", x)) +
  geom_smooth(method = "lm", se = T, color = "orange") +
  geom_jitter(alpha = 0.12, width = 0.2) +
  theme_classic() +
  stat_cor(method = "spearman", cor.coef.name = "rho") +
  theme(axis.title.y = element_blank())
ggsave("old graphs/gene-expression-ratios-CAZ.pdf", height = 8, width =10)
```








```{r}
DNAs = colnames(DNA_RNA)[grepl("DNA_", colnames(DNA_RNA))]
RNAs = colnames(DNA_RNA)[grepl("RNA_", colnames(DNA_RNA))]

tmp = cbind(DNA_RNA$Ocean_DNA, rowSums(DNA_RNA[DNAs]), rowSums(DNA_RNA[RNAs]))
```


```{r}
gen_plot <- function(gene_name, log_scale, percent_scale, upscale, 
                     DNA_color, RNA_color, txt_top, txt_bot){
  log_scale <- log_scale + upscale
  DNA_gene <- paste("log_dna",gene_name, sep="_")
  RNA_gene <- paste("log_rna",gene_name, sep="_")
  tmp1 <- DNA_RNA[,c(DNA_gene, "Layer_RNA")]
  tmp2 <- DNA_RNA[,c(RNA_gene, "Layer_RNA")]
  tmp1$type <- "DNA"
  tmp2$type <- "RNA"
  
  tmp1 <- filter_outliers(tmp1, DNA_gene)
  tmp2 <- filter_outliers(tmp2, RNA_gene)
  
  g_cols <- c("log_abundance", "depth", "type")
  DNA_RNA_bar <- rbind(`colnames<-`(tmp1, g_cols),
                       `colnames<-`(tmp2, g_cols))
  
  sum_bar <- DNA_RNA_bar %>% 
    group_by(type, depth) %>% 
    summarise(median_abun = median(log_abundance) + upscale, 
    mean_abun = mean(log_abundance),
    lower = quantile(log_abundance, probs = c(.25)) + upscale, 
    upper = quantile(log_abundance, probs = c(.75)) + upscale)
  
  x_max <- max(max(DNA_RNA[,c(DNA_gene)]), max(DNA_RNA[,c(RNA_gene)]))+ upscale
  labD = "DNA"
  labR = "RNA"
  
  p <- ggplot(sum_bar, aes(y=fct_rev(depth), x=median_abun, fill=fct_rev(type))) + 
    geom_bar(stat="identity", position=position_dodge()) +
    # scale_x_continuous(breaks = log_scale, labels = percent_scale) +# 
    geom_errorbar(aes(xmin=lower, xmax=upper), width=.2,
                  position=position_dodge(.9)) + 
    scale_fill_manual(values = c("DNA" = DNA_color, "RNA" = RNA_color))+
    annotate(geom="text", x=x_max*txt_top, y=3.25, label=labD) + 
    annotate(geom="text", x=x_max*txt_bot, y=2.8, label=labR) + 
    theme_classic() +
    
    theme(axis.title.y = element_blank(),
          legend.position = "none")
  
  if (gene_name == "trans") {
    return(p + xlab("% reads mapped to transposases"))
  } else if (gene_name == "defense") {
    return(p + xlab("% reads mapped to defense mechanism ORFs"))
  } else if (gene_name == "biofilm"){
    p <- p + xlab("% reads mapped to biofilm ORFs")
    p <- p + scale_y_discrete(labels=c("MES  \n(n = 22)",
                                       "DCM  \n(n = 36)",
                                       "SRF   \n(n = 64)"))
  }
  return(p)
}
```

## Using the function
```{r}
exp_biofilm <- gen_plot("biofilm", c(-4.301, -4, -3.698, -3.3979),
                        c("0.005%", "0.01%", "0.02%", "0.04%"), 4.3,
                        "gray", "green", 0.54, 0.44)

exp_trans <- gen_plot("trans", c(-4.523, -4, -3.523, -3, -2.523), 
                      c("0.003%", "0.01%", "0.03%", "0.10%","0.3%"),
                      4.52, "gray", "light blue", 0.54, 0.46)

exp_defense <- gen_plot("defense", c(-3, -2.699, -2.398, -2.097),
                        c("0.1%", "0.2%", "0.4%","0.8%"),
                        3, "gray", "red", 0.1, 0.1)

gen_plot("signalT", c(-3, -2.699, -2.398, -2.097),
         c("0.1%", "0.2%", "0.4%","0.8%"),
         3, "gray", "red", 0.1, 0.1)

gen_plot("coenzyme", c(-3, -2.699, -2.398, -2.097),
         c("0.1%", "0.2%", "0.4%","0.8%"),
         3, "gray", "red", 0.1, 0.1)

gen_plot("energyPC", c(-3, -2.699, -2.398, -2.097),
         c("0.1%", "0.2%", "0.4%","0.8%"),
         3, "gray", "red", 0.1, 0.1)

gen_plot("replication", c(-3, -2.699, -2.398, -2.097),
         c("0.1%", "0.2%", "0.4%","0.8%"),
         3, "gray", "red", 0.1, 0.1)
```


## Marking ranks are not too useful
```{r}
pre.melt = function(DNA_RNA, in_cols) {
  target = DNA_RNA[c("connector_DNA", in_cols)]
  target$depth = factor(DNA_RNA$Layer_RNA,
     levels = c("MES","DCM","SRF"))
  target.melt = target %>%
    reshape2::melt(id.var = c("depth", "connector_DNA"))
  return(target.melt)
}

base.col = c("depth", "connector_DNA", "gene")

exp.melt = pre.melt(DNA_RNA, exp_col)
colnames(exp.melt) = c(base.col, "expression")
exp.melt$gene = gsub(" exp", "",gsub("_"," ", exp.melt$gene))


abun.melt = pre.melt(DNA_RNA, dna_col)
colnames(abun.melt) = c(base.col, "abundance")
abun.melt$gene = gsub("_", " ", gsub("DNA_", "", abun.melt$gene))
abun.melt = abun.melt %>%
  group_by(gene) %>%
  mutate(group_rank = order(abundance, decreasing=TRUE)) %>%
  mutate(group_rank = as.numeric(group_rank))

exp.stats = exp.melt %>%
  group_by(gene) %>%
  wilcox_test(expression~depth) %>%
  adjust_pvalue(method = "BH") %>%
  add_significance() %>% 
  add_xy_position() %>%
  mutate(y.position = log2(y.position))

exp.abun.melt = merge(exp.melt, abun.melt, by = base.col)

ggplot(exp.abun.melt, aes(y = group_rank, x = depth)) +
  facet_wrap(~gene) +
  # scale_y_continuous(trans = "log2",
  #   labels = function(x) ifelse(x == 0, "0", x)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(aes(color = group_rank), size = 1, alpha = 0.5) +
  scale_color_gradient2(low="blue", mid="white", high="red",
                        midpoint = 63) +
  stat_pvalue_manual(exp.stats, tip.length = 0.01,
    label = "p.adj.signif", hide.ns = TRUE ) +
  coord_flip() +
  theme_classic() +
  theme(axis.title.y = element_blank())

```








sel_cols = c("biofilm_exp_rate", "defense_exp_rate", "trans_exp_rate", "Layer_RNA")
melted <- DNA_RNA[, sel_cols] %>%
  `colnames<-`(c("Biofilm", "Defense mech.", "Transposase", "Depth")) %>%
  reshape2::melt(id.vars=c('Depth'),var='Gene')

library(ggpubr)
library(rstatix)
melted$Depth <- fct_rev(melted$Depth)
melted$log_exp <- log10(melted$value)

# three gene's expression looks very normal with log transform
# hist((filter(melted, Gene == "Biofilm"))$log_exp) 

stat.test <- melted %>%
  group_by(Gene) %>%
  t_test(log_exp~Depth) %>%
  add_significance("p") %>% 
  add_xy_position()
stat.test2 <- filter(stat.test, group1 =="MES")
stat.test2$y.position <- log10(c(3.2, 3.8, 1.9, 1.2, 5.6, 25)) # bio, def, trans

exp_summary <- melted %>% # 
  ggplot(aes(x=Depth, y=value, fill = Gene)) +
  scale_fill_manual(values = c("green", "red","light blue"))+
  facet_wrap(~ Gene, ncol = 3, scales = "free_x") +
  geom_boxplot(outlier.colour = "gray") +
  stat_pvalue_manual(stat.test2, label = "p.signif", tip.length = 0.01) +
  ylab("RNA/DNA ratio (%RNA / %DNA per sample)") +
  scale_y_log10() +
  theme_classic() +
  theme(legend.position = "none",
        axis.title.y=element_blank()) +
  coord_flip() 

ggarrange(nrow = 2, ncol = 1,
  ggarrange(exp_biofilm, exp_defense, widths = c(0.5,0.46), labels = c("A","B"), ncol = 2),
  ggarrange(exp_trans, exp_summary, widths = c(0.4,0.6), labels = c("C","D"), ncol = 2))

ggsave("F7_DNA-RNA-ratio_target-genes.png", 
       plot = last_plot(),
       height = 4.3,
       width = 7.5)
ggsave("F7_DNA-RNA-ratio_target-genes.pdf", 
       plot = last_plot(),
       height = 4.3,
       width = 7.5)


