---
title: "Graph metagenomic particle association and transposase abundance"
author: "Jimmy Zhong"
date: "8/19/2022"
output: html_document
---

```{r setup, include=FALSE}
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source("./init_share.R")
g(DNA_tara, RNA_tara, DNA_RNA_tara) %=% init_tara()
mala_cov <- init_mala_cov()
mala_cov$size_fraction <- ifelse(mala_cov$lower_filter_size == "0.2", "planktonic", "particle")
DNA_tara$size_fraction <- ifelse(DNA_tara$upper_size_dna == "1.6", "planktonic", "particle")
```

## Data preparations
```{r}
DNA_RNA_tara_g = DNA_RNA_tara %>%
  filter(Layer_RNA != "MIX" & Ocean != "Arctic Ocean") %>%
  mutate(DNA_trans_percent = DNA_transposase * 100)

RNA_tara_g = RNA_tara %>%
  filter(Layer_RNA != "MIX" & Ocean != "Arctic Ocean") %>%
  mutate(RNA_trans_percent = RNA_transposase * 100) %>%
  mutate(Depth = factor(Layer_RNA, levels = c("BAT","MES","DCM","SRF")))

pt_sel_col <- c("log_dna_trans", "log_dna_defense","DNA_transposase",
                "log_dna_sect_CAZ","log_dna_sect_pep",
                "DNA_CAZyme","DNA_peptidase", "DNA_biofilm",
                "percent_dna_sect_CAZpep","percent_dna_sect_CAZ",
                "percent_dna_sect_pep",
                "Ocean_DNA", "Layer_DNA", "Oxygen_DNA", "size_fraction")

tmp = rbind(mala_cov[ ,pt_sel_col], DNA_tara[,pt_sel_col])
pt_to_graph = tmp %>%
  mutate(DNA_trans_percent = DNA_transposase * 100) %>%
  filter(Layer_DNA != "MIX" & Ocean_DNA != "Arctic") %>%
  mutate(Depth = factor(Layer_DNA, levels = c("SRF","DCM","MES","BAT")))

pt_to_graph$log_dna_biofilm = log2(pt_to_graph$DNA_biofilm)
RNA_tara_g$log_rna_biofilm = log2(RNA_tara_g$RNA_biofilm)
biofilm_dna_scale = c(0.005, 0.01, 0.02, 0.04, 0.08)
biofilm_rna_scale = c(0.001, 0.003, 0.01, 0.03, 0.1)
```

## DNA particle assocation to transpoase abundance
```{r}
id.col = c("Depth", "Ocean_DNA", "DNA_trans_percent")
pt.depth.melt = pt_to_graph[c(id.col,
                              "percent_dna_sect_CAZ",
                              "percent_dna_sect_pep")] %>%
  reshape2::melt(id = id.col)

DNA.depth = pt.depth.melt %>%
  mutate(gene = ifelse(variable == "percent_dna_sect_CAZ", "CAZyme", "peptidase")) %>%
  ggplot(aes(x=value, y=DNA_trans_percent)) +
  scale_color_brewer(palette="Set1")+
  facet_grid(Depth ~ gene, scale = "free_x") +
  geom_point(aes(color = Ocean_DNA)) + 
  stat_cor(method = "spearman", cor.coef.name = "rho",
           aes(label = paste(..r.label.., cut(..p.., 
    breaks = c(-Inf, 0.0001, 0.001, 0.01, 0.05, Inf),
    labels = c("'****'", "'***'", "'**'", "'*'", "'ns'")), 
    sep = "~"))) +
  theme_classic() +
  scale_y_log10() +
  labs( color = "Ocean", 
    y = "transposase gene abundance (%)",
    x = "secretory CAZyme/peptidase genes (%)") +
  geom_smooth(method = "lm", se = T)  +
  theme(axis.text.x = element_text(color="black"),
        axis.text.y = element_text(color="black"),
        legend.position = "None",
        legend.title = element_blank())
```


```{r}
RNA.id = c("Layer_RNA", "Ocean", "RNA_trans_percent")
RNA.melt = RNA_tara_g[c(RNA.id, "percent_rna_sect_CAZ",
                                "percent_rna_sect_pep")] %>%
  reshape2::melt(id = RNA.id)

RNA.melt$gene = ifelse(RNA.melt$variable == "percent_rna_sect_CAZ", "CAZyme", "peptidase")

RNA.depth = RNA.melt %>%
  ggplot(aes(x=value, y=RNA_trans_percent)) +
  scale_color_brewer(palette="Set1")+
  facet_grid(Layer_RNA ~ gene, scale = "free_x") +
  # facet_wrap(~Layer_RNA, scale = "free_x") +
  geom_point(aes(color = Ocean)) + 
  stat_cor(method = "spearman", cor.coef.name = "rho",
           aes(label = paste(..r.label.., cut(..p.., 
    breaks = c(-Inf, 0.0001, 0.001, 0.01, 0.05, Inf),
    labels = c("'****'", "'***'", "'**'", "'*'", "'ns'")), 
    sep = "~"))) +
  theme_classic() +
  scale_y_log10() +
  scale_x_log10() +
  labs( color = "Ocean", 
    y = "transposase transcript abundance (%)",
    x = "secretory CAZyme/peptidase transcript (%)\n\n\nMetatranscriptome samples unavailable\nin the bathypelagic zone (BAT)") +
  geom_smooth(method = "lm", se = T)  +
  theme(axis.text.x = element_text(color="black"),
        axis.text.y = element_text(color="black"),
        legend.position = "bottom",
        legend.title = element_blank()) +
  guides(color=guide_legend(nrow=3, byrow=TRUE))

ggarrange(DNA.depth, RNA.depth)
## supplementary figure 3
ggsave("out/sectCAZpep_transpose_DNARNA.pdf",
       plot = last_plot(), width = 9, height = 8)
```

## Biofilm and transposase abundance correlation (DNA and RNA)
```{r}
trans_dna_scale = c(0.001, 0.003, 0.01, 0.03, 0.1, 0.3)

bt_DNA = pt_to_graph %>%
  ggplot(aes(x=log_dna_biofilm, y=log_dna_trans)) +
  scale_x_continuous(breaks = log2(biofilm_dna_scale/100), labels = biofilm_dna_scale) +
  scale_y_continuous(breaks = log2(trans_dna_scale/100), labels = trans_dna_scale) +
  scale_color_brewer(palette="Set1")+
  facet_wrap(~Depth) +
  geom_point(aes(color = Ocean_DNA), size = 1) + 
  theme_classic() +
  stat_cor(method = "spearman", cor.coef.name = "rho",
           aes(label = paste(..r.label.., cut(..p.., 
    breaks = c(-Inf, 0.0001, 0.001, 0.01, 0.05, Inf),
    labels = c("'****'", "'***'", "'**'", "'*'", "'ns'")), 
    sep = "~"))) +
  xlab("biofilm-associated ORFs gene abundance (%)") +
  ylab("transposase gene abundance (%)") +
  geom_smooth(method = "lm")  +
  theme(legend.position = "none") # legend.position = "none"

ann_text <- data.frame(log_rna_biofilm = -13, log_rna_trans = -12,
            label = "Metatranscriptomes\nnot available",
            Layer_RNA = factor("BAT",levels = c("SRF", "DCM", "MES", "BAT")))

biofilm_rna_scale1 = c(0.001, 0.003, 0.01, 0.03, 0.1, 0.3)
trans_rna_scale = c(0.001, 0.003, 0.01, 0.03, 0.1, 0.3)
bt_RNA = RNA_tara_g %>%
  filter(!is.na(Layer_RNA)) %>%
  mutate(Layer_RNA = factor(Layer_RNA, levels = c("SRF", "DCM", "MES", "BAT"))) %>% # for the BAT text
  ggplot(aes(x=log_rna_biofilm, y=log_rna_trans)) +
  theme_classic() +
  scale_y_continuous(breaks = log2(biofilm_rna_scale1/100), labels = biofilm_rna_scale1) +
  scale_x_continuous(breaks = log2(trans_rna_scale/100), labels = trans_rna_scale) +
  facet_wrap(~Layer_RNA, nrow = 2, drop = FALSE) +
  geom_point(aes(color = Ocean), size = 1) + 
  scale_color_brewer(palette="Set1")+
  stat_cor(method = "spearman", cor.coef.name = "rho",
           aes(label = paste(..r.label.., cut(..p.., 
    breaks = c(-Inf, 0.0001, 0.001, 0.01, 0.05, Inf),
    labels = c("'****'", "'***'", "'**'", "'*'", "'ns'")), 
    sep = "~"))) +
  xlab("biofilm-associated ORFs transcript abundance (%)") +
  ylab("transposase transcript abundance (%)") +
  geom_smooth(method = "lm") +
  geom_text(data = ann_text,label=ann_text$label)

ggarrange(bt_DNA, bt_RNA, labels = c("A", "B"), widths = c(0.42,0.6), ncol =2)

# supplementary figure 4
ggsave("out/biofilm_transpose_DNARNA.pdf",
       plot = last_plot(), width = 10, height = 4.5)
```


## indiviudal CAZyme and peptidase

Correlation with transposase abundance (DNA and RNA)
```{r}
gen_corr_shared = function(g, gene, type) {
  g = g + theme_classic() +
    theme(axis.text.x = element_text(color="black"),
          axis.text.y = element_text(color="black")) +
    scale_y_continuous(trans = "log10",
                       limits = c(min(RNA_tara_g$RNA_trans_percent)*0.99,
                                  1.2),
                       labels = function(x) ifelse(x == 0, "0", x)) +
    scale_color_brewer(palette="Set1") +
    # geom_smooth(method = "lm", se = T) +
    stat_cor(method = "spearman", cor.coef.name = "rho") +
    labs( color = "Depth",
          x = sprintf("secretory %s %ss (%%)", gene, type),
          y = sprintf("transposase %s abundance (%%)", type))
  return(g)
}

gen_RNA_corr = function(x_var, gene, type){
  g = RNA_tara_g %>%
  ggplot(aes_string(x=x_var, y="RNA_trans_percent")) +
    geom_point(aes(color = Layer_RNA)) + 
    scale_x_log10()
  g = gen_corr_shared(g, gene, type)
  return(g)
}

gen_DNA_corr = function(x_var, gene, type){
  g = pt_to_graph %>%
    ggplot(aes_string(x=x_var, y="DNA_trans_percent")) +
    geom_point(aes(color = Depth))
  g = gen_corr_shared(g, gene, type)
  return(g)
}

CAZ_RNA = gen_RNA_corr("percent_rna_sect_CAZ", "CAZyme", "transcript") +
  theme(legend.position = "none")

pep_RNA = gen_RNA_corr("percent_rna_sect_pep", "peptidase", "transcript") +
  theme( axis.title.y = element_blank() )

CAZ_DNA = gen_DNA_corr("percent_dna_sect_CAZ", "CAZyme", "gene") +
  theme(legend.position = "none")

pep_DNA = gen_DNA_corr("percent_dna_sect_pep", "peptidase", "gene") +
  theme( axis.title.y = element_blank() )

ggarrange(CAZ_DNA, pep_DNA, CAZ_RNA, pep_RNA,
          # labels = c("A", "B", "C", "D"),
          ncol = 2, nrow = 2, widths = c(0.45, 0.55))

ggsave(paste0(googleDrive.loc, "/Figure2_CAZyme_peptidase_separate_corr.pdf"),
       plot = last_plot(), width = 7, height = 5.6)

ggsave("out/CAZyme_peptidase_separate_corr.pdf",
       width = 7, height = 5.6)
```

## Show that deep oceans rely more on particle-associated lifestyle (RNA)
```{r}
# RNA_tara_g %>%
#   ggplot(aes(x = percent_rna_sect_pep, y= log_rna_trans)) +
#   geom_point(aes(color = Layer_RNA)) +
#   scale_x_log10() +
#   facet_wrap(~Ocean) +
#   geom_smooth(method = "lm", se = F)

summary(lm(log_dna_trans~Depth + Ocean + Oxygen_DNA, pt_to_graph))
summary(lm(log_dna_trans~Depth + Ocean + Oxygen_DNA, DNA_RNA_tara))

gen_depth_shared = function(data, data.test) {
  g = ggplot(data, aes(x=Depth, y=cur_var)) + 
    geom_boxplot(outlier.colour = NA) +
    geom_jitter(aes(color=Ocean), alpha = 0.7,
                width = 0.2, height = 0, size = 1) +
    stat_pvalue_manual(data.test, label = "p.adj.signif", 
                       tip.length = 0.01, hide.ns = TRUE)+ 
    scale_color_brewer(palette="Set1") +
    theme_classic() + 
    theme(axis.title.y = element_blank(),
          legend.title = element_blank(),
          axis.text.x = element_text(color="black"),
          axis.text.y = element_text(color="black")) +
    coord_flip()
  return(g)
}

gen_RNA_depth = function(x_var, position_vec, 
  text_pos, BAT_text = "NA"){
  RNA_tara_g['cur_var'] = RNA_tara_g[x_var]
  secretory_rna.test <- RNA_tara_g %>%
  rstatix::wilcox_test(cur_var ~ Layer_RNA) %>%
  adjust_pvalue(method = "BH") %>%
  add_significance() %>%
  add_xy_position() %>%
  mutate(xmin = 5 - xmin, xmax = 5 - xmax) # the empty BAT group and the coor_flip makes everything weird, but I think this is right. 1 -> SRF: 5 - 1 = 4 (top, new SRF); 3 -> MES: 5 - 3 = 2 (second to the last, MES)
  
  secretory_rna.test$y.position = position_vec
  
  ann_text2 = data.frame(cur_var = text_pos,
    label = BAT_text,
    Depth = factor("BAT",levels = c("BAT","MES","DCM","SRF")))
  
  set.seed(33)

  RNA_particle = gen_depth_shared(RNA_tara_g, secretory_rna.test) +
    scale_x_discrete(drop=FALSE) +
    geom_text(data = ann_text2,label=ann_text2$label)
  return(RNA_particle)
}

# Maybe I don't need this?
# gen_RNA_depth("percent_rna_sect_CAZpep", c(31, 33.3, 29.7), 21) +
#   ylab("secretory CAZyme and peptidase transcript (%)")

CAZ_RNA_depth = gen_RNA_depth(
  x_var = "percent_rna_sect_CAZ",
  position_vec = c(13, 16, 14.5),
  text_pos = 11,
  BAT_text = "Metatranscriptome samples not available") +
  ylab("secretory CAZyme transcript (%)") +
  theme(legend.position = "none")

pep_RNA_depth = gen_RNA_depth("percent_rna_sect_pep", c(47, 53, 51), 32) +
  ylab("secretory peptidase transcript (%)") +
  theme(legend.position = "none")
```

## Show that deep oceans rely more on particle-associated lifestyle (DNA)
```{r}
gen_DNA_depth = function(x_var, position_vec, seed = 33){
  pt_to_graph['cur_var'] = pt_to_graph[x_var]
  secretory_dna.test = pt_to_graph %>%
  mutate(Depth = fct_rev(Depth)) %>% 
  rstatix::wilcox_test(cur_var ~ Depth) %>%
  adjust_pvalue(method = "BH") %>%
  add_significance() %>%
  add_xy_position()

  secretory_dna.test$y.position = position_vec

  set.seed(seed)
  DNA_particle = pt_to_graph %>%
    mutate(Depth = fct_rev(Depth),
           Ocean = Ocean_DNA) %>% 
    gen_depth_shared(secretory_dna.test)
  return(DNA_particle)
}

CAZ_DNA_depth = gen_DNA_depth("percent_dna_sect_CAZ",
  c(21, 22, 23, 19, 20, 10.0), seed = 44) +
  ylab("secretory CAZyme gene (%)")

pep_DNA_depth = gen_DNA_depth("percent_dna_sect_pep",
  c(38, 39, 40, 35, 36, 10)) +
  ylab("secretory peptidase gene (%)") +
  theme(legend.position = "none")
```

## Adding biofilm gene and transcript abundance with depth
```{r}
biofilm_DNA_depth = gen_DNA_depth("log_dna_biofilm",
                                  seed = 32,
                                  c(0, -7.2, -7.1, -7.5, -7.4, 0)) +
  ylab("biofilm-associated gene abundance (%)") +
  theme(legend.position = "none") +
  scale_y_continuous(breaks = log2(biofilm_dna_scale/100), labels = biofilm_dna_scale)

biofilm_RNA_depth = gen_RNA_depth("log_rna_biofilm",
  position_vec = c(0, -6.6, -6.8), 
  text_pos = -8.75) +
  ylab("biofilm-associated transcript abundance (%)") +
  theme(legend.position = "none") +
  scale_y_continuous(breaks = log2(biofilm_rna_scale/100), labels = biofilm_rna_scale)


blank_plot = ggplot() + theme_void()

ggarrange( nrow = 3,
  ggarrange( CAZ_RNA_depth, CAZ_DNA_depth, ncol = 2,
        labels = c("A","B"), widths = c(0.4, 0.55)),
  ggarrange( pep_RNA_depth, pep_DNA_depth, blank_plot, ncol = 3,
        labels = c("C","D"), widths = c(0.4, 0.4, 0.15)),
  ggarrange( biofilm_RNA_depth, biofilm_DNA_depth, blank_plot, ncol = 3,
        labels = c("E","F"), widths = c(0.4, 0.4, 0.15)))

## supplementary figure 2
ggsave("out/sectCAZpep_increase_depth.pdf",
       plot = last_plot(), width = 9, height = 7)
```


## stats section for particle-assocation transposase correlation

The anova for RNA is insignificant: 0.8823
But it is stronger for the DNA: 9.231e-11

> get_r(DNA.corr.particle) - get_r(DNA.corr.depth)
[1] 0.06233815
> get_r(DNA.corr.particle)
[1] 0.7128721
> get_r(DNA.corr.depth)
[1] 0.6505339

```{r}
reg.stats = function(big_df, depth, feature, trans) {
  tmp_df = big_df[c(feature, depth, "Ocean", trans)]
  colnames(tmp_df) = c("feature", "depth", "ocean", "trans")
  corr.depth = lm(trans~depth, tmp_df)
  corr.particle = lm(trans~depth + feature, tmp_df)
  print(summary(corr.particle))
  ano_obj = anova(corr.depth, corr.particle)
  
  n_row = data.frame(feature = feature,
     sample.n = nrow(tmp_df),
     depth.norm = shapiro.test(resid(corr.depth))$p.val,
     feature.norm = shapiro.test(resid(corr.particle))$p.val,
     depth.R = get_r(corr.depth), 
     feature.R = get_r(corr.particle),
     p.val = ano_obj$`Pr(>F)`[2])
  return(n_row)
}

cols = c("log_dna_trans","Depth","Ocean_DNA", "Layer_DNA")
tmp = rbind(DNA_tara[cols], mala_cov[cols])
tmp = tmp[tmp$Ocean_DNA != "Arctic", ]

summary(lm(log_dna_trans~Layer_DNA, tmp ))
summary(lm(log_dna_trans~Layer_DNA, pt_to_graph))

# DNA_RNA_tara_g$Depth = DNA_RNA_tara_g$Layer_RNA
# DNA_RNA_tara_g$log_percent_CAZ = log(DNA_RNA_tara_g$percent_rna_sect_CAZ)
pt_to_graph$Ocean = pt_to_graph$Ocean_DNA
RNA_tara_g$log_percent_CAZ = log(RNA_tara_g$percent_rna_sect_CAZ)
RNA_tara_g$log_percent_pep = log(RNA_tara_g$percent_rna_sect_pep)
reg1 = reg.stats(pt_to_graph[!is.na(pt_to_graph$Oxygen_DNA), ],
                 "Layer_DNA", "Oxygen_DNA", "log_dna_trans")
reg2 = reg.stats(pt_to_graph, "Layer_DNA", "percent_dna_sect_CAZ", "log_dna_trans")
reg3 = reg.stats(pt_to_graph, "Layer_DNA", "percent_dna_sect_pep", "log_dna_trans")
# RNA_tara.oxy = RNA_tara_g[!is.na(RNA_tara_g$Oxygen_RNA), ]
# reg4 = reg.stats(RNA_tara.oxy, "Layer_RNA", "Oxygen_RNA", "log_rna_trans")
reg5 = reg.stats(RNA_tara_g, "Layer_RNA", "log_percent_CAZ", "log_rna_trans")
reg6 = reg.stats(RNA_tara_g, "Layer_RNA", "log_percent_pep", "log_rna_trans")

CAZpep.reg = rbind(reg1, reg2, reg3, reg4, reg5, reg6) %>%
  adjust_pvalue(method = "BY")

write.csv(CAZpep.reg, file = "redux_out/regression-results.csv")

# ggplot(RNA_tara.oxy, aes(x = Oxygen_RNA, y = log_rna_trans)) +
#   geom_point() +
#   facet_wrap(~Layer_RNA)
# cor.test(RNA_tara.oxy$Oxygen_RNA, RNA_tara.oxy$log_rna_trans, method = "sp")
```


## Arctic Ocean have higher secretory CAZyme peptidase percentage

Have to do log transformation to make sure that both groups have
normal distributions.

RNA p-val = 1.037e-09
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -0.3011970 -0.1680382
sample estimates:
mean of x mean of y 
 2.862235  3.096852 
 
** the difference is around 26.4% in RNA **
> exp(3.096852-2.862235)
[1] 1.264424

** 95% CI ** 
> exp(0.3011970)
[1] 1.351476
> exp(0.1680382)
[1] 1.182982


```{r}
RNA.arctic = RNA_tara_g[c("percent_rna_sect_CAZpep", "Ocean")] %>%
  mutate(is_arctic = ifelse(Ocean == "Arctic Ocean", "Arctic", "Non-arctic"))

RNA.arctic = log(RNA_tara_g[RNA_tara_g$Ocean == "Arctic Ocean",
                        "percent_rna_sect_CAZpep"])
RNA.non.arctic = log(RNA_tara_g[RNA_tara_g$Ocean != "Arctic Ocean",
                        "percent_rna_sect_CAZpep"])

# normality assumption met
shapiro.test(RNA.non.arctic)
shapiro.test(RNA.arctic)
# equal variance not met
var.test(RNA.non.arctic, RNA.arctic)
# do t test
t.test(RNA.non.arctic, RNA.arctic, var.equal = FALSE)
```

## Now do the same thing for DNA

Can't get the normality assumption to work...
Go with non parametric

```{r}
DNA.arctic = pt_to_graph[pt_to_graph$Ocean_DNA == "Arctic",]$percent_dna_sect_CAZpep
DNA.non.arctic = pt_to_graph[pt_to_graph$Ocean_DNA != "Arctic",]$percent_dna_sect_CAZpep

# normality assumption met
shapiro.test(DNA.arctic)
shapiro.test(DNA.non.arctic)

wilcox.test(DNA.arctic, DNA.non.arctic)
wilcox.test(RNA.arctic, RNA.non.arctic)
```


## Will worry about scales later
```{r}
trans_RNA_scale <- c(-4.5, -4, -3.5, -3)
trans_RNA_percent_scale <- c("0.003", "0.010", "0.032", "0.100")
trans_scale <- c(-2.5, -3, -3.5, -4, -4.5)
trans_percent_scale <- c("0.316", "0.100", "0.032", "0.01", "0.003")
CAZ_percent_scale <- c(0.8, 0.4, 0.2, 0.1)
CAZ_scale <- log10(CAZ_percent_scale/100)

# BAT.sample.size = mala_cov[c('sample', 'lower_filter_size')]
# BAT.sample.size$filter_size = ifelse(
#   BAT.sample.size$lower_filter_size == 0.2, "0.2-0.8", "0.8-5.0")
# 
# BAT.sample.size = filter(BAT.sample.size, filter_size == "0.2-0.8")
# 
# BAT.COG.corr = COG.corr[COG.corr$Layer_DNA == "BAT", ]
# BAT.COG.corr = merge(BAT.COG.corr, BAT.sample.size, on = "sample")
# tara.COG.corr = COG.corr[COG.corr$Layer_DNA != "BAT", ]
```


> summary(pt_to_graph$DNA_peptidase)
    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
0.004958 0.026718 0.029942 0.029231 0.034359 0.046223 
> summary(pt_to_graph$DNA_CAZyme)
    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
0.004601 0.015382 0.018450 0.018156 0.020963 0.029557
> sd(pt_to_graph$DNA_CAZyme)
[1] 0.004086007
> sd(pt_to_graph$DNA_peptidase)
[1] 0.00777566

The total abundance of CAZyme and peptidase are similar in magnitude

## RNA particle association to transposase abundance
```{r}
ann_text = data.frame(percent_rna_sect_CAZ = 20, DNA_trans_percent = 0.05,
           label = "No metatranscriptome\nsamples available",
           Layer_RNA = factor("BAT",levels = c("SRF", "DCM", "MES", "BAT")))

RNA_tara_g %>%
  ggplot(aes(x=percent_rna_sect_pep, y=RNA_transposase)) +
  scale_color_brewer(palette="Set1")+
  stat_cor(method = "spearman", cor.coef.name = "rho") +
  facet_wrap(~Layer_RNA, ncol = 2) +
  geom_point(aes(color = Ocean)) + 
  scale_y_log10() +
  theme_classic() +
  labs( y = "Transposase gene abundance (%)",
        x = "Secretory CAZyme and peptidase transcript (%)") +
  geom_smooth(method = "lm", se = T)
  # geom_text(data = ann_text,label=ann_text$label)
  # theme(legend.position = "none")

## supplementary figure 1
# ggsave("out/sectCAZpep_transpose_RNA.pdf",
#        plot = last_plot(), width = 6.5, height = 5) 
```